// canvas-sketch/dist/canvas-sketch.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
  typeof define === 'function' && define.amd ? define(factory) :
  (global = global || self, global.canvasSketch = factory());
})(this, (function () {
  function _defineProperties(e, r) {
    for (var t = 0; t < r.length; t++) {
      var o = r[t];
      o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o);
    }
  }
  function _createClass(e, r, t) {
    return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", {
      writable: !1
    }), e;
  }
  function _extends() {
    return _extends = Object.assign ? Object.assign.bind() : function (n) {
      for (var e = 1; e < arguments.length; e++) {
        var t = arguments[e];
        for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]);
      }
      return n;
    }, _extends.apply(null, arguments);
  }
  function _toPrimitive(t, r) {
    if ("object" != typeof t || !t) return t;
    var e = t[Symbol.toPrimitive];
    if (void 0 !== e) {
      var i = e.call(t, r || "default");
      if ("object" != typeof i) return i;
      throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === r ? String : Number)(t);
  }
  function _toPropertyKey(t) {
    var i = _toPrimitive(t, "string");
    return "symbol" == typeof i ? i : i + "";
  }

  var commonjsGlobal = typeof globalThis !== 'undefined' ? globalThis : typeof window !== 'undefined' ? window : typeof global !== 'undefined' ? global : typeof self !== 'undefined' ? self : {};

  function createCommonjsModule(fn) {
    var module = { exports: {} };
  	return fn(module, module.exports), module.exports;
  }

  var browser =
    commonjsGlobal.performance &&
    commonjsGlobal.performance.now ? function now() {
      return performance.now()
    } : Date.now || function now() {
      return +new Date
    };

  var isPromise_1 = isPromise;
  var _default = isPromise;

  function isPromise(obj) {
    return !!obj && (typeof obj === 'object' || typeof obj === 'function') && typeof obj.then === 'function';
  }
  isPromise_1.default = _default;

  var isObject = function isObject(x) {
  	return typeof x === 'object' && x !== null;
  };

  var isWindow = function (obj) {

    if (obj == null) {
      return false;
    }

    var o = Object(obj);

    return o === o.window;
  };

  function isNode (val) {
    if (!isObject(val) || !isWindow(window) || typeof window.Node !== 'function') {
      return false
    }

    return typeof val.nodeType === 'number' &&
      typeof val.nodeName === 'string'
  }

  var isDom = isNode;

  // TODO: We can remove a huge chunk of bundle size by using a smaller
  function getClientAPI() {
    return typeof window !== "undefined" && window["canvas-sketch-cli"];
  }
  function defined$1() {
    for (var i = 0; i < arguments.length; i++) {
      if (arguments[i] != null) {
        return arguments[i];
      }
    }
    return undefined;
  }
  function isBrowser() {
    return typeof document !== "undefined";
  }
  function isWebGLContext(ctx) {
    return ctx && typeof ctx.clear === "function" && typeof ctx.clearColor === "function" && typeof ctx.bufferData === "function";
  }
  function is2DContext(ctx) {
    return ctx && typeof ctx.save === "function" && typeof ctx.scale === "function" && typeof ctx.restore === "function";
  }
  function isCanvas(element) {
    return isDom(element) && /canvas/i.test(element.nodeName) && typeof element.getContext === "function";
  }

  var toStr$4 = Object.prototype.toString;

  var isArguments$2 = function isArguments(value) {
  	var str = toStr$4.call(value);
  	var isArgs = str === '[object Arguments]';
  	if (!isArgs) {
  		isArgs = str !== '[object Array]' &&
  			value !== null &&
  			typeof value === 'object' &&
  			typeof value.length === 'number' &&
  			value.length >= 0 &&
  			toStr$4.call(value.callee) === '[object Function]';
  	}
  	return isArgs;
  };

  var keysShim$1;
  if (!Object.keys) {
  	// modified from https://github.com/es-shims/es5-shim
  	var has = Object.prototype.hasOwnProperty;
  	var toStr$3 = Object.prototype.toString;
  	var isArgs = isArguments$2; // eslint-disable-line global-require
  	var isEnumerable = Object.prototype.propertyIsEnumerable;
  	var hasDontEnumBug = !isEnumerable.call({ toString: null }, 'toString');
  	var hasProtoEnumBug = isEnumerable.call(function () {}, 'prototype');
  	var dontEnums = [
  		'toString',
  		'toLocaleString',
  		'valueOf',
  		'hasOwnProperty',
  		'isPrototypeOf',
  		'propertyIsEnumerable',
  		'constructor'
  	];
  	var equalsConstructorPrototype = function (o) {
  		var ctor = o.constructor;
  		return ctor && ctor.prototype === o;
  	};
  	var excludedKeys = {
  		$applicationCache: true,
  		$console: true,
  		$external: true,
  		$frame: true,
  		$frameElement: true,
  		$frames: true,
  		$innerHeight: true,
  		$innerWidth: true,
  		$onmozfullscreenchange: true,
  		$onmozfullscreenerror: true,
  		$outerHeight: true,
  		$outerWidth: true,
  		$pageXOffset: true,
  		$pageYOffset: true,
  		$parent: true,
  		$scrollLeft: true,
  		$scrollTop: true,
  		$scrollX: true,
  		$scrollY: true,
  		$self: true,
  		$webkitIndexedDB: true,
  		$webkitStorageInfo: true,
  		$window: true
  	};
  	var hasAutomationEqualityBug = (function () {
  		/* global window */
  		if (typeof window === 'undefined') { return false; }
  		for (var k in window) {
  			try {
  				if (!excludedKeys['$' + k] && has.call(window, k) && window[k] !== null && typeof window[k] === 'object') {
  					try {
  						equalsConstructorPrototype(window[k]);
  					} catch (e) {
  						return true;
  					}
  				}
  			} catch (e) {
  				return true;
  			}
  		}
  		return false;
  	}());
  	var equalsConstructorPrototypeIfNotBuggy = function (o) {
  		/* global window */
  		if (typeof window === 'undefined' || !hasAutomationEqualityBug) {
  			return equalsConstructorPrototype(o);
  		}
  		try {
  			return equalsConstructorPrototype(o);
  		} catch (e) {
  			return false;
  		}
  	};

  	keysShim$1 = function keys(object) {
  		var isObject = object !== null && typeof object === 'object';
  		var isFunction = toStr$3.call(object) === '[object Function]';
  		var isArguments = isArgs(object);
  		var isString = isObject && toStr$3.call(object) === '[object String]';
  		var theKeys = [];

  		if (!isObject && !isFunction && !isArguments) {
  			throw new TypeError('Object.keys called on a non-object');
  		}

  		var skipProto = hasProtoEnumBug && isFunction;
  		if (isString && object.length > 0 && !has.call(object, 0)) {
  			for (var i = 0; i < object.length; ++i) {
  				theKeys.push(String(i));
  			}
  		}

  		if (isArguments && object.length > 0) {
  			for (var j = 0; j < object.length; ++j) {
  				theKeys.push(String(j));
  			}
  		} else {
  			for (var name in object) {
  				if (!(skipProto && name === 'prototype') && has.call(object, name)) {
  					theKeys.push(String(name));
  				}
  			}
  		}

  		if (hasDontEnumBug) {
  			var skipConstructor = equalsConstructorPrototypeIfNotBuggy(object);

  			for (var k = 0; k < dontEnums.length; ++k) {
  				if (!(skipConstructor && dontEnums[k] === 'constructor') && has.call(object, dontEnums[k])) {
  					theKeys.push(dontEnums[k]);
  				}
  			}
  		}
  		return theKeys;
  	};
  }
  var implementation$3 = keysShim$1;

  var slice = Array.prototype.slice;


  var origKeys = Object.keys;
  var keysShim = origKeys ? function keys(o) { return origKeys(o); } : implementation$3;

  var originalKeys = Object.keys;

  keysShim.shim = function shimObjectKeys() {
  	if (Object.keys) {
  		var keysWorksWithArguments = (function () {
  			// Safari 5.0 bug
  			var args = Object.keys(arguments);
  			return args && args.length === arguments.length;
  		}(1, 2));
  		if (!keysWorksWithArguments) {
  			Object.keys = function keys(object) { // eslint-disable-line func-name-matching
  				if (isArguments$2(object)) {
  					return originalKeys(slice.call(object));
  				}
  				return originalKeys(object);
  			};
  		}
  	} else {
  		Object.keys = keysShim;
  	}
  	return Object.keys || keysShim;
  };

  var objectKeys = keysShim;

  /** @type {import('./shams')} */
  /* eslint complexity: [2, 18], max-statements: [2, 33] */
  var shams$1 = function hasSymbols() {
  	if (typeof Symbol !== 'function' || typeof Object.getOwnPropertySymbols !== 'function') { return false; }
  	if (typeof Symbol.iterator === 'symbol') { return true; }

  	/** @type {{ [k in symbol]?: unknown }} */
  	var obj = {};
  	var sym = Symbol('test');
  	var symObj = Object(sym);
  	if (typeof sym === 'string') { return false; }

  	if (Object.prototype.toString.call(sym) !== '[object Symbol]') { return false; }
  	if (Object.prototype.toString.call(symObj) !== '[object Symbol]') { return false; }

  	// temp disabled per https://github.com/ljharb/object.assign/issues/17
  	// if (sym instanceof Symbol) { return false; }
  	// temp disabled per https://github.com/WebReflection/get-own-property-symbols/issues/4
  	// if (!(symObj instanceof Symbol)) { return false; }

  	// if (typeof Symbol.prototype.toString !== 'function') { return false; }
  	// if (String(sym) !== Symbol.prototype.toString.call(sym)) { return false; }

  	var symVal = 42;
  	obj[sym] = symVal;
  	for (var _ in obj) { return false; } // eslint-disable-line no-restricted-syntax, no-unreachable-loop
  	if (typeof Object.keys === 'function' && Object.keys(obj).length !== 0) { return false; }

  	if (typeof Object.getOwnPropertyNames === 'function' && Object.getOwnPropertyNames(obj).length !== 0) { return false; }

  	var syms = Object.getOwnPropertySymbols(obj);
  	if (syms.length !== 1 || syms[0] !== sym) { return false; }

  	if (!Object.prototype.propertyIsEnumerable.call(obj, sym)) { return false; }

  	if (typeof Object.getOwnPropertyDescriptor === 'function') {
  		// eslint-disable-next-line no-extra-parens
  		var descriptor = /** @type {PropertyDescriptor} */ (Object.getOwnPropertyDescriptor(obj, sym));
  		if (descriptor.value !== symVal || descriptor.enumerable !== true) { return false; }
  	}

  	return true;
  };

  /** @type {import('.')} */
  var shams = function hasToStringTagShams() {
  	return shams$1() && !!Symbol.toStringTag;
  };

  /** @type {import('.')} */
  var esObjectAtoms = Object;

  /** @type {import('.')} */
  var esErrors = Error;

  /** @type {import('./eval')} */
  var _eval = EvalError;

  /** @type {import('./range')} */
  var range = RangeError;

  /** @type {import('./ref')} */
  var ref = ReferenceError;

  /** @type {import('./syntax')} */
  var syntax = SyntaxError;

  /** @type {import('./type')} */
  var type = TypeError;

  /** @type {import('./uri')} */
  var uri = URIError;

  /** @type {import('./abs')} */
  var abs$1 = Math.abs;

  /** @type {import('./floor')} */
  var floor$1 = Math.floor;

  /** @type {import('./max')} */
  var max$2 = Math.max;

  /** @type {import('./min')} */
  var min$1 = Math.min;

  /** @type {import('./pow')} */
  var pow$1 = Math.pow;

  /** @type {import('./round')} */
  var round$2 = Math.round;

  /** @type {import('./isNaN')} */
  var _isNaN = Number.isNaN || function isNaN(a) {
  	return a !== a;
  };

  var $isNaN = _isNaN;

  /** @type {import('./sign')} */
  var sign$1 = function sign(number) {
  	if ($isNaN(number) || number === 0) {
  		return number;
  	}
  	return number < 0 ? -1 : +1;
  };

  /** @type {import('./gOPD')} */
  var gOPD$2 = Object.getOwnPropertyDescriptor;

  var require$$0$1 = gOPD$2;

  /** @type {import('.')} */
  var $gOPD$1 = require$$0$1;

  if ($gOPD$1) {
  	try {
  		$gOPD$1([], 'length');
  	} catch (e) {
  		// IE 8 has a broken gOPD
  		$gOPD$1 = null;
  	}
  }

  var gopd = $gOPD$1;

  /** @type {import('.')} */
  var $defineProperty$1 = Object.defineProperty || false;
  if ($defineProperty$1) {
  	try {
  		$defineProperty$1({}, 'a', { value: 1 });
  	} catch (e) {
  		// IE 8 has a broken defineProperty
  		$defineProperty$1 = false;
  	}
  }

  var esDefineProperty = $defineProperty$1;

  var origSymbol = typeof Symbol !== 'undefined' && Symbol;


  /** @type {import('.')} */
  var hasSymbols$2 = function hasNativeSymbols() {
  	if (typeof origSymbol !== 'function') { return false; }
  	if (typeof Symbol !== 'function') { return false; }
  	if (typeof origSymbol('foo') !== 'symbol') { return false; }
  	if (typeof Symbol('bar') !== 'symbol') { return false; }

  	return shams$1();
  };

  /** @type {import('./Reflect.getPrototypeOf')} */
  var Reflect_getPrototypeOf = (typeof Reflect !== 'undefined' && Reflect.getPrototypeOf) || null;

  var $Object$2 = esObjectAtoms;

  /** @type {import('./Object.getPrototypeOf')} */
  var Object_getPrototypeOf = $Object$2.getPrototypeOf || null;

  /* eslint no-invalid-this: 1 */

  var ERROR_MESSAGE = 'Function.prototype.bind called on incompatible ';
  var toStr$2 = Object.prototype.toString;
  var max$1 = Math.max;
  var funcType = '[object Function]';

  var concatty = function concatty(a, b) {
      var arr = [];

      for (var i = 0; i < a.length; i += 1) {
          arr[i] = a[i];
      }
      for (var j = 0; j < b.length; j += 1) {
          arr[j + a.length] = b[j];
      }

      return arr;
  };

  var slicy = function slicy(arrLike, offset) {
      var arr = [];
      for (var i = offset || 0, j = 0; i < arrLike.length; i += 1, j += 1) {
          arr[j] = arrLike[i];
      }
      return arr;
  };

  var joiny = function (arr, joiner) {
      var str = '';
      for (var i = 0; i < arr.length; i += 1) {
          str += arr[i];
          if (i + 1 < arr.length) {
              str += joiner;
          }
      }
      return str;
  };

  var implementation$2 = function bind(that) {
      var target = this;
      if (typeof target !== 'function' || toStr$2.apply(target) !== funcType) {
          throw new TypeError(ERROR_MESSAGE + target);
      }
      var args = slicy(arguments, 1);

      var bound;
      var binder = function () {
          if (this instanceof bound) {
              var result = target.apply(
                  this,
                  concatty(args, arguments)
              );
              if (Object(result) === result) {
                  return result;
              }
              return this;
          }
          return target.apply(
              that,
              concatty(args, arguments)
          );

      };

      var boundLength = max$1(0, target.length - args.length);
      var boundArgs = [];
      for (var i = 0; i < boundLength; i++) {
          boundArgs[i] = '$' + i;
      }

      bound = Function('binder', 'return function (' + joiny(boundArgs, ',') + '){ return binder.apply(this,arguments); }')(binder);

      if (target.prototype) {
          var Empty = function Empty() {};
          Empty.prototype = target.prototype;
          bound.prototype = new Empty();
          Empty.prototype = null;
      }

      return bound;
  };

  var functionBind = Function.prototype.bind || implementation$2;

  /** @type {import('./functionCall')} */
  var functionCall = Function.prototype.call;

  /** @type {import('./functionApply')} */
  var functionApply = Function.prototype.apply;

  /** @type {import('./reflectApply')} */
  var reflectApply = typeof Reflect !== 'undefined' && Reflect && Reflect.apply;

  /** @type {import('./actualApply')} */
  var actualApply = reflectApply || functionBind.call(functionCall, functionApply);

  var $TypeError$1 = type;

  /** @type {(args: [Function, thisArg?: unknown, ...args: unknown[]]) => Function} TODO FIXME, find a way to use import('.') */
  var callBindApplyHelpers = function callBindBasic(args) {
  	if (args.length < 1 || typeof args[0] !== 'function') {
  		throw new $TypeError$1('a function is required');
  	}
  	return actualApply(functionBind, functionCall, args);
  };

  var gOPD$1 = gopd;

  var hasProtoAccessor;
  try {
  	// eslint-disable-next-line no-extra-parens, no-proto
  	hasProtoAccessor = /** @type {{ __proto__?: typeof Array.prototype }} */ ([]).__proto__ === Array.prototype;
  } catch (e) {
  	if (!e || typeof e !== 'object' || !('code' in e) || e.code !== 'ERR_PROTO_ACCESS') {
  		throw e;
  	}
  }

  // eslint-disable-next-line no-extra-parens
  var desc = !!hasProtoAccessor && gOPD$1 && gOPD$1(Object.prototype, /** @type {keyof typeof Object.prototype} */ ('__proto__'));

  var $Object$1 = Object;
  var $getPrototypeOf = $Object$1.getPrototypeOf;

  /** @type {import('./get')} */
  var get = desc && typeof desc.get === 'function'
  	? callBindApplyHelpers([desc.get])
  	: typeof $getPrototypeOf === 'function'
  		? /** @type {import('./get')} */ function getDunder(value) {
  			// eslint-disable-next-line eqeqeq
  			return $getPrototypeOf(value == null ? value : $Object$1(value));
  		}
  		: false;

  var getDunderProto = get;

  /** @type {import('.')} */
  var getProto = Reflect_getPrototypeOf
  	? function getProto(O) {
  		// @ts-expect-error TS can't narrow inside a closure, for some reason
  		return Reflect_getPrototypeOf(O);
  	}
  	: Object_getPrototypeOf
  		? function getProto(O) {
  			if (!O || (typeof O !== 'object' && typeof O !== 'function')) {
  				throw new TypeError('getProto: not an object');
  			}
  			// @ts-expect-error TS can't narrow inside a closure, for some reason
  			return Object_getPrototypeOf(O);
  		}
  		: getDunderProto
  			? function getProto(O) {
  				// @ts-expect-error TS can't narrow inside a closure, for some reason
  				return getDunderProto(O);
  			}
  			: null;

  var call = Function.prototype.call;
  var $hasOwn = Object.prototype.hasOwnProperty;


  /** @type {import('.')} */
  var hasown = functionBind.call(call, $hasOwn);

  var $TypeError = esErrors;

  var $EvalError = _eval;

  var $RangeError = range;

  var $ReferenceError = ref;

  var $SyntaxError = syntax;

  var $URIError = uri;

  var abs = abs$1;

  var floor = floor$1;

  var max = max$2;

  var min = min$1;

  var pow = pow$1;

  var round$1 = round$2;

  var sign = sign$1;

  var $defineProperty = esDefineProperty;

  var hasOwn = hasown;

  var undefined$1;



















  var $Function = Function;

  // eslint-disable-next-line consistent-return
  var getEvalledConstructor = function (expressionSyntax) {
  	try {
  		return $Function('"use strict"; return (' + expressionSyntax + ').constructor;')();
  	} catch (e) {}
  };




  var throwTypeError = function () {
  	throw new $TypeError$1();
  };
  var ThrowTypeError = gOPD$1
  	? (function () {
  		try {
  			// eslint-disable-next-line no-unused-expressions, no-caller, no-restricted-properties
  			arguments.callee; // IE 8 does not throw here
  			return throwTypeError;
  		} catch (calleeThrows) {
  			try {
  				// IE 8 throws on Object.getOwnPropertyDescriptor(arguments, '')
  				return gOPD$1(arguments, 'callee').get;
  			} catch (gOPDthrows) {
  				return throwTypeError;
  			}
  		}
  	}())
  	: throwTypeError;

  var hasSymbols$1 = hasSymbols$2();








  var needsEval = {};

  var TypedArray = typeof Uint8Array === 'undefined' || !getProto ? undefined$1 : getProto(Uint8Array);

  var INTRINSICS = {
  	__proto__: null,
  	'%AggregateError%': typeof AggregateError === 'undefined' ? undefined$1 : AggregateError,
  	'%Array%': Array,
  	'%ArrayBuffer%': typeof ArrayBuffer === 'undefined' ? undefined$1 : ArrayBuffer,
  	'%ArrayIteratorPrototype%': hasSymbols$1 && getProto ? getProto([][Symbol.iterator]()) : undefined$1,
  	'%AsyncFromSyncIteratorPrototype%': undefined$1,
  	'%AsyncFunction%': needsEval,
  	'%AsyncGenerator%': needsEval,
  	'%AsyncGeneratorFunction%': needsEval,
  	'%AsyncIteratorPrototype%': needsEval,
  	'%Atomics%': typeof Atomics === 'undefined' ? undefined$1 : Atomics,
  	'%BigInt%': typeof BigInt === 'undefined' ? undefined$1 : BigInt,
  	'%BigInt64Array%': typeof BigInt64Array === 'undefined' ? undefined$1 : BigInt64Array,
  	'%BigUint64Array%': typeof BigUint64Array === 'undefined' ? undefined$1 : BigUint64Array,
  	'%Boolean%': Boolean,
  	'%DataView%': typeof DataView === 'undefined' ? undefined$1 : DataView,
  	'%Date%': Date,
  	'%decodeURI%': decodeURI,
  	'%decodeURIComponent%': decodeURIComponent,
  	'%encodeURI%': encodeURI,
  	'%encodeURIComponent%': encodeURIComponent,
  	'%Error%': $TypeError,
  	'%eval%': eval, // eslint-disable-line no-eval
  	'%EvalError%': $EvalError,
  	'%Float16Array%': typeof Float16Array === 'undefined' ? undefined$1 : Float16Array,
  	'%Float32Array%': typeof Float32Array === 'undefined' ? undefined$1 : Float32Array,
  	'%Float64Array%': typeof Float64Array === 'undefined' ? undefined$1 : Float64Array,
  	'%FinalizationRegistry%': typeof FinalizationRegistry === 'undefined' ? undefined$1 : FinalizationRegistry,
  	'%Function%': $Function,
  	'%GeneratorFunction%': needsEval,
  	'%Int8Array%': typeof Int8Array === 'undefined' ? undefined$1 : Int8Array,
  	'%Int16Array%': typeof Int16Array === 'undefined' ? undefined$1 : Int16Array,
  	'%Int32Array%': typeof Int32Array === 'undefined' ? undefined$1 : Int32Array,
  	'%isFinite%': isFinite,
  	'%isNaN%': isNaN,
  	'%IteratorPrototype%': hasSymbols$1 && getProto ? getProto(getProto([][Symbol.iterator]())) : undefined$1,
  	'%JSON%': typeof JSON === 'object' ? JSON : undefined$1,
  	'%Map%': typeof Map === 'undefined' ? undefined$1 : Map,
  	'%MapIteratorPrototype%': typeof Map === 'undefined' || !hasSymbols$1 || !getProto ? undefined$1 : getProto(new Map()[Symbol.iterator]()),
  	'%Math%': Math,
  	'%Number%': Number,
  	'%Object%': $Object$2,
  	'%Object.getOwnPropertyDescriptor%': gOPD$1,
  	'%parseFloat%': parseFloat,
  	'%parseInt%': parseInt,
  	'%Promise%': typeof Promise === 'undefined' ? undefined$1 : Promise,
  	'%Proxy%': typeof Proxy === 'undefined' ? undefined$1 : Proxy,
  	'%RangeError%': $RangeError,
  	'%ReferenceError%': $ReferenceError,
  	'%Reflect%': typeof Reflect === 'undefined' ? undefined$1 : Reflect,
  	'%RegExp%': RegExp,
  	'%Set%': typeof Set === 'undefined' ? undefined$1 : Set,
  	'%SetIteratorPrototype%': typeof Set === 'undefined' || !hasSymbols$1 || !getProto ? undefined$1 : getProto(new Set()[Symbol.iterator]()),
  	'%SharedArrayBuffer%': typeof SharedArrayBuffer === 'undefined' ? undefined$1 : SharedArrayBuffer,
  	'%String%': String,
  	'%StringIteratorPrototype%': hasSymbols$1 && getProto ? getProto(''[Symbol.iterator]()) : undefined$1,
  	'%Symbol%': hasSymbols$1 ? Symbol : undefined$1,
  	'%SyntaxError%': $SyntaxError,
  	'%ThrowTypeError%': ThrowTypeError,
  	'%TypedArray%': TypedArray,
  	'%TypeError%': $TypeError$1,
  	'%Uint8Array%': typeof Uint8Array === 'undefined' ? undefined$1 : Uint8Array,
  	'%Uint8ClampedArray%': typeof Uint8ClampedArray === 'undefined' ? undefined$1 : Uint8ClampedArray,
  	'%Uint16Array%': typeof Uint16Array === 'undefined' ? undefined$1 : Uint16Array,
  	'%Uint32Array%': typeof Uint32Array === 'undefined' ? undefined$1 : Uint32Array,
  	'%URIError%': $URIError,
  	'%WeakMap%': typeof WeakMap === 'undefined' ? undefined$1 : WeakMap,
  	'%WeakRef%': typeof WeakRef === 'undefined' ? undefined$1 : WeakRef,
  	'%WeakSet%': typeof WeakSet === 'undefined' ? undefined$1 : WeakSet,

  	'%Function.prototype.call%': functionCall,
  	'%Function.prototype.apply%': functionApply,
  	'%Object.defineProperty%': $defineProperty,
  	'%Object.getPrototypeOf%': Object_getPrototypeOf,
  	'%Math.abs%': abs,
  	'%Math.floor%': floor,
  	'%Math.max%': max,
  	'%Math.min%': min,
  	'%Math.pow%': pow,
  	'%Math.round%': round$1,
  	'%Math.sign%': sign,
  	'%Reflect.getPrototypeOf%': Reflect_getPrototypeOf
  };

  if (getProto) {
  	try {
  		null.error; // eslint-disable-line no-unused-expressions
  	} catch (e) {
  		// https://github.com/tc39/proposal-shadowrealm/pull/384#issuecomment-1364264229
  		var errorProto = getProto(getProto(e));
  		INTRINSICS['%Error.prototype%'] = errorProto;
  	}
  }

  var doEval = function doEval(name) {
  	var value;
  	if (name === '%AsyncFunction%') {
  		value = getEvalledConstructor('async function () {}');
  	} else if (name === '%GeneratorFunction%') {
  		value = getEvalledConstructor('function* () {}');
  	} else if (name === '%AsyncGeneratorFunction%') {
  		value = getEvalledConstructor('async function* () {}');
  	} else if (name === '%AsyncGenerator%') {
  		var fn = doEval('%AsyncGeneratorFunction%');
  		if (fn) {
  			value = fn.prototype;
  		}
  	} else if (name === '%AsyncIteratorPrototype%') {
  		var gen = doEval('%AsyncGenerator%');
  		if (gen && getProto) {
  			value = getProto(gen.prototype);
  		}
  	}

  	INTRINSICS[name] = value;

  	return value;
  };

  var LEGACY_ALIASES = {
  	__proto__: null,
  	'%ArrayBufferPrototype%': ['ArrayBuffer', 'prototype'],
  	'%ArrayPrototype%': ['Array', 'prototype'],
  	'%ArrayProto_entries%': ['Array', 'prototype', 'entries'],
  	'%ArrayProto_forEach%': ['Array', 'prototype', 'forEach'],
  	'%ArrayProto_keys%': ['Array', 'prototype', 'keys'],
  	'%ArrayProto_values%': ['Array', 'prototype', 'values'],
  	'%AsyncFunctionPrototype%': ['AsyncFunction', 'prototype'],
  	'%AsyncGenerator%': ['AsyncGeneratorFunction', 'prototype'],
  	'%AsyncGeneratorPrototype%': ['AsyncGeneratorFunction', 'prototype', 'prototype'],
  	'%BooleanPrototype%': ['Boolean', 'prototype'],
  	'%DataViewPrototype%': ['DataView', 'prototype'],
  	'%DatePrototype%': ['Date', 'prototype'],
  	'%ErrorPrototype%': ['Error', 'prototype'],
  	'%EvalErrorPrototype%': ['EvalError', 'prototype'],
  	'%Float32ArrayPrototype%': ['Float32Array', 'prototype'],
  	'%Float64ArrayPrototype%': ['Float64Array', 'prototype'],
  	'%FunctionPrototype%': ['Function', 'prototype'],
  	'%Generator%': ['GeneratorFunction', 'prototype'],
  	'%GeneratorPrototype%': ['GeneratorFunction', 'prototype', 'prototype'],
  	'%Int8ArrayPrototype%': ['Int8Array', 'prototype'],
  	'%Int16ArrayPrototype%': ['Int16Array', 'prototype'],
  	'%Int32ArrayPrototype%': ['Int32Array', 'prototype'],
  	'%JSONParse%': ['JSON', 'parse'],
  	'%JSONStringify%': ['JSON', 'stringify'],
  	'%MapPrototype%': ['Map', 'prototype'],
  	'%NumberPrototype%': ['Number', 'prototype'],
  	'%ObjectPrototype%': ['Object', 'prototype'],
  	'%ObjProto_toString%': ['Object', 'prototype', 'toString'],
  	'%ObjProto_valueOf%': ['Object', 'prototype', 'valueOf'],
  	'%PromisePrototype%': ['Promise', 'prototype'],
  	'%PromiseProto_then%': ['Promise', 'prototype', 'then'],
  	'%Promise_all%': ['Promise', 'all'],
  	'%Promise_reject%': ['Promise', 'reject'],
  	'%Promise_resolve%': ['Promise', 'resolve'],
  	'%RangeErrorPrototype%': ['RangeError', 'prototype'],
  	'%ReferenceErrorPrototype%': ['ReferenceError', 'prototype'],
  	'%RegExpPrototype%': ['RegExp', 'prototype'],
  	'%SetPrototype%': ['Set', 'prototype'],
  	'%SharedArrayBufferPrototype%': ['SharedArrayBuffer', 'prototype'],
  	'%StringPrototype%': ['String', 'prototype'],
  	'%SymbolPrototype%': ['Symbol', 'prototype'],
  	'%SyntaxErrorPrototype%': ['SyntaxError', 'prototype'],
  	'%TypedArrayPrototype%': ['TypedArray', 'prototype'],
  	'%TypeErrorPrototype%': ['TypeError', 'prototype'],
  	'%Uint8ArrayPrototype%': ['Uint8Array', 'prototype'],
  	'%Uint8ClampedArrayPrototype%': ['Uint8ClampedArray', 'prototype'],
  	'%Uint16ArrayPrototype%': ['Uint16Array', 'prototype'],
  	'%Uint32ArrayPrototype%': ['Uint32Array', 'prototype'],
  	'%URIErrorPrototype%': ['URIError', 'prototype'],
  	'%WeakMapPrototype%': ['WeakMap', 'prototype'],
  	'%WeakSetPrototype%': ['WeakSet', 'prototype']
  };



  var $concat = functionBind.call(functionCall, Array.prototype.concat);
  var $spliceApply = functionBind.call(functionApply, Array.prototype.splice);
  var $replace = functionBind.call(functionCall, String.prototype.replace);
  var $strSlice = functionBind.call(functionCall, String.prototype.slice);
  var $exec$1 = functionBind.call(functionCall, RegExp.prototype.exec);

  /* adapted from https://github.com/lodash/lodash/blob/4.17.15/dist/lodash.js#L6735-L6744 */
  var rePropName = /[^%.[\]]+|\[(?:(-?\d+(?:\.\d+)?)|(["'])((?:(?!\2)[^\\]|\\.)*?)\2)\]|(?=(?:\.|\[\])(?:\.|\[\]|%$))/g;
  var reEscapeChar = /\\(\\)?/g; /** Used to match backslashes in property paths. */
  var stringToPath = function stringToPath(string) {
  	var first = $strSlice(string, 0, 1);
  	var last = $strSlice(string, -1);
  	if (first === '%' && last !== '%') {
  		throw new $SyntaxError('invalid intrinsic syntax, expected closing `%`');
  	} else if (last === '%' && first !== '%') {
  		throw new $SyntaxError('invalid intrinsic syntax, expected opening `%`');
  	}
  	var result = [];
  	$replace(string, rePropName, function (match, number, quote, subString) {
  		result[result.length] = quote ? $replace(subString, reEscapeChar, '$1') : number || match;
  	});
  	return result;
  };
  /* end adaptation */

  var getBaseIntrinsic = function getBaseIntrinsic(name, allowMissing) {
  	var intrinsicName = name;
  	var alias;
  	if (hasOwn(LEGACY_ALIASES, intrinsicName)) {
  		alias = LEGACY_ALIASES[intrinsicName];
  		intrinsicName = '%' + alias[0] + '%';
  	}

  	if (hasOwn(INTRINSICS, intrinsicName)) {
  		var value = INTRINSICS[intrinsicName];
  		if (value === needsEval) {
  			value = doEval(intrinsicName);
  		}
  		if (typeof value === 'undefined' && !allowMissing) {
  			throw new $TypeError$1('intrinsic ' + name + ' exists, but is not available. Please file an issue!');
  		}

  		return {
  			alias: alias,
  			name: intrinsicName,
  			value: value
  		};
  	}

  	throw new $SyntaxError('intrinsic ' + name + ' does not exist!');
  };

  var getIntrinsic = function GetIntrinsic(name, allowMissing) {
  	if (typeof name !== 'string' || name.length === 0) {
  		throw new $TypeError$1('intrinsic name must be a non-empty string');
  	}
  	if (arguments.length > 1 && typeof allowMissing !== 'boolean') {
  		throw new $TypeError$1('"allowMissing" argument must be a boolean');
  	}

  	if ($exec$1(/^%?[^%]*%?$/, name) === null) {
  		throw new $SyntaxError('`%` may not be present anywhere but at the beginning and end of the intrinsic name');
  	}
  	var parts = stringToPath(name);
  	var intrinsicBaseName = parts.length > 0 ? parts[0] : '';

  	var intrinsic = getBaseIntrinsic('%' + intrinsicBaseName + '%', allowMissing);
  	var intrinsicRealName = intrinsic.name;
  	var value = intrinsic.value;
  	var skipFurtherCaching = false;

  	var alias = intrinsic.alias;
  	if (alias) {
  		intrinsicBaseName = alias[0];
  		$spliceApply(parts, $concat([0, 1], alias));
  	}

  	for (var i = 1, isOwn = true; i < parts.length; i += 1) {
  		var part = parts[i];
  		var first = $strSlice(part, 0, 1);
  		var last = $strSlice(part, -1);
  		if (
  			(
  				(first === '"' || first === "'" || first === '`')
  				|| (last === '"' || last === "'" || last === '`')
  			)
  			&& first !== last
  		) {
  			throw new $SyntaxError('property names with quotes must have matching quotes');
  		}
  		if (part === 'constructor' || !isOwn) {
  			skipFurtherCaching = true;
  		}

  		intrinsicBaseName += '.' + part;
  		intrinsicRealName = '%' + intrinsicBaseName + '%';

  		if (hasOwn(INTRINSICS, intrinsicRealName)) {
  			value = INTRINSICS[intrinsicRealName];
  		} else if (value != null) {
  			if (!(part in value)) {
  				if (!allowMissing) {
  					throw new $TypeError$1('base intrinsic for ' + name + ' exists, but the property is not available.');
  				}
  				return void undefined$1;
  			}
  			if (gOPD$1 && (i + 1) >= parts.length) {
  				var desc = gOPD$1(value, part);
  				isOwn = !!desc;

  				// By convention, when a data property is converted to an accessor
  				// property to emulate a data property that does not suffer from
  				// the override mistake, that accessor's getter is marked with
  				// an `originalValue` property. Here, when we detect this, we
  				// uphold the illusion by pretending to see that original data
  				// property, i.e., returning the value rather than the getter
  				// itself.
  				if (isOwn && 'get' in desc && !('originalValue' in desc.get)) {
  					value = desc.get;
  				} else {
  					value = value[part];
  				}
  			} else {
  				isOwn = hasOwn(value, part);
  				value = value[part];
  			}

  			if (isOwn && !skipFurtherCaching) {
  				INTRINSICS[intrinsicRealName] = value;
  			}
  		}
  	}
  	return value;
  };

  var GetIntrinsic = getIntrinsic;

  /** @type {(thisArg: string, searchString: string, position?: number) => number} */
  var $indexOf = callBindApplyHelpers([GetIntrinsic('%String.prototype.indexOf%')]);

  /** @type {import('.')} */
  var callBound$1 = function callBoundIntrinsic(name, allowMissing) {
  	/* eslint no-extra-parens: 0 */

  	var intrinsic = /** @type {(this: unknown, ...args: unknown[]) => unknown} */ (GetIntrinsic(name, !!allowMissing));
  	if (typeof intrinsic === 'function' && $indexOf(name, '.prototype.') > -1) {
  		return callBindApplyHelpers(/** @type {const} */ ([intrinsic]));
  	}
  	return intrinsic;
  };

  var callBound = callBound$1;

  var hasToStringTag$2 = shams();


  var $toString$1 = callBound('Object.prototype.toString');

  /** @type {import('.')} */
  var isStandardArguments = function isArguments(value) {
  	if (
  		hasToStringTag$2
  		&& value
  		&& typeof value === 'object'
  		&& Symbol.toStringTag in value
  	) {
  		return false;
  	}
  	return $toString$1(value) === '[object Arguments]';
  };

  /** @type {import('.')} */
  var isLegacyArguments = function isArguments(value) {
  	if (isStandardArguments(value)) {
  		return true;
  	}
  	return value !== null
  		&& typeof value === 'object'
  		&& 'length' in value
  		&& typeof value.length === 'number'
  		&& value.length >= 0
  		&& $toString$1(value) !== '[object Array]'
  		&& 'callee' in value
  		&& $toString$1(value.callee) === '[object Function]';
  };

  var supportsStandardArguments = (function () {
  	return isStandardArguments(arguments);
  }());

  // @ts-expect-error TODO make this not error
  isStandardArguments.isLegacyArguments = isLegacyArguments; // for tests

  /** @type {import('.')} */
  var isArguments$1 = supportsStandardArguments ? isStandardArguments : isLegacyArguments;

  /** @type {import('.')} */
  var defineDataProperty = function defineDataProperty(
  	obj,
  	property,
  	value
  ) {
  	if (!obj || (typeof obj !== 'object' && typeof obj !== 'function')) {
  		throw new $TypeError$1('`obj` must be an object or a function`');
  	}
  	if (typeof property !== 'string' && typeof property !== 'symbol') {
  		throw new $TypeError$1('`property` must be a string or a symbol`');
  	}
  	if (arguments.length > 3 && typeof arguments[3] !== 'boolean' && arguments[3] !== null) {
  		throw new $TypeError$1('`nonEnumerable`, if provided, must be a boolean or null');
  	}
  	if (arguments.length > 4 && typeof arguments[4] !== 'boolean' && arguments[4] !== null) {
  		throw new $TypeError$1('`nonWritable`, if provided, must be a boolean or null');
  	}
  	if (arguments.length > 5 && typeof arguments[5] !== 'boolean' && arguments[5] !== null) {
  		throw new $TypeError$1('`nonConfigurable`, if provided, must be a boolean or null');
  	}
  	if (arguments.length > 6 && typeof arguments[6] !== 'boolean') {
  		throw new $TypeError$1('`loose`, if provided, must be a boolean');
  	}

  	var nonEnumerable = arguments.length > 3 ? arguments[3] : null;
  	var nonWritable = arguments.length > 4 ? arguments[4] : null;
  	var nonConfigurable = arguments.length > 5 ? arguments[5] : null;
  	var loose = arguments.length > 6 ? arguments[6] : false;

  	/* @type {false | TypedPropertyDescriptor<unknown>} */
  	var desc = !!gOPD$1 && gOPD$1(obj, property);

  	if ($defineProperty) {
  		$defineProperty(obj, property, {
  			configurable: nonConfigurable === null && desc ? desc.configurable : !nonConfigurable,
  			enumerable: nonEnumerable === null && desc ? desc.enumerable : !nonEnumerable,
  			value: value,
  			writable: nonWritable === null && desc ? desc.writable : !nonWritable
  		});
  	} else if (loose || (!nonEnumerable && !nonWritable && !nonConfigurable)) {
  		// must fall back to [[Set]], and was not explicitly asked to make non-enumerable, non-writable, or non-configurable
  		obj[property] = value; // eslint-disable-line no-param-reassign
  	} else {
  		throw new $SyntaxError('This environment does not support defining a property as non-configurable, non-writable, or non-enumerable.');
  	}
  };

  var hasPropertyDescriptors = function hasPropertyDescriptors() {
  	return !!$defineProperty;
  };

  hasPropertyDescriptors.hasArrayLengthDefineBug = function hasArrayLengthDefineBug() {
  	// node v0.6 has a bug where array lengths can be Set but not Defined
  	if (!$defineProperty) {
  		return null;
  	}
  	try {
  		return $defineProperty([], 'length', { value: 1 }).length !== 1;
  	} catch (e) {
  		// In Firefox 4-22, defining length on an array throws an exception.
  		return true;
  	}
  };

  var hasPropertyDescriptors_1 = hasPropertyDescriptors;

  var define = defineDataProperty;

  var require$$0 = hasPropertyDescriptors_1;

  var hasSymbols = typeof Symbol === 'function' && typeof Symbol('foo') === 'symbol';

  var toStr$1 = Object.prototype.toString;
  var concat = Array.prototype.concat;


  var isFunction = function (fn) {
  	return typeof fn === 'function' && toStr$1.call(fn) === '[object Function]';
  };

  var supportsDescriptors$2 = require$$0();

  var defineProperty$1 = function (object, name, value, predicate) {
  	if (name in object) {
  		if (predicate === true) {
  			if (object[name] === value) {
  				return;
  			}
  		} else if (!isFunction(predicate) || !predicate()) {
  			return;
  		}
  	}

  	if (supportsDescriptors$2) {
  		define(object, name, value, true);
  	} else {
  		define(object, name, value);
  	}
  };

  var defineProperties = function (object, map) {
  	var predicates = arguments.length > 2 ? arguments[2] : {};
  	var props = objectKeys(map);
  	if (hasSymbols) {
  		props = concat.call(props, Object.getOwnPropertySymbols(map));
  	}
  	for (var i = 0; i < props.length; i += 1) {
  		defineProperty$1(object, props[i], map[props[i]], predicates[props[i]]);
  	}
  };

  defineProperties.supportsDescriptors = !!supportsDescriptors$2;

  var defineProperties_1 = defineProperties;

  var hasDescriptors$1 = require$$0();



  var $floor = GetIntrinsic('%Math.floor%');

  /** @type {import('.')} */
  var setFunctionLength$1 = function setFunctionLength(fn, length) {
  	if (typeof fn !== 'function') {
  		throw new $TypeError$1('`fn` is not a function');
  	}
  	if (typeof length !== 'number' || length < 0 || length > 0xFFFFFFFF || $floor(length) !== length) {
  		throw new $TypeError$1('`length` must be a positive 32-bit integer');
  	}

  	var loose = arguments.length > 2 && !!arguments[2];

  	var functionLengthIsConfigurable = true;
  	var functionLengthIsWritable = true;
  	if ('length' in fn && gOPD$1) {
  		var desc = gOPD$1(fn, 'length');
  		if (desc && !desc.configurable) {
  			functionLengthIsConfigurable = false;
  		}
  		if (desc && !desc.writable) {
  			functionLengthIsWritable = false;
  		}
  	}

  	if (functionLengthIsConfigurable || functionLengthIsWritable || !loose) {
  		if (hasDescriptors$1) {
  			define(/** @type {Parameters<define>[0]} */ (fn), 'length', length, true, true);
  		} else {
  			define(/** @type {Parameters<define>[0]} */ (fn), 'length', length);
  		}
  	}
  	return fn;
  };

  /** @type {import('./applyBind')} */
  var applyBind = function applyBind() {
  	return actualApply(functionBind, functionApply, arguments);
  };

  var setFunctionLength = setFunctionLength$1;

  var callBind = createCommonjsModule(function (module) {








  module.exports = function callBind(originalFunction) {
  	var func = callBindApplyHelpers(arguments);
  	var adjustedLength = 1 + originalFunction.length - (arguments.length - 1);
  	return setFunctionLength(
  		func,
  		adjustedLength > 0 ? adjustedLength : 0,
  		true
  	);
  };

  if ($defineProperty) {
  	$defineProperty(module.exports, 'apply', { value: applyBind });
  } else {
  	module.exports.apply = applyBind;
  }
  });

  var numberIsNaN = function (value) {
  	return value !== value;
  };

  var implementation$1 = function is(a, b) {
  	if (a === 0 && b === 0) {
  		return 1 / a === 1 / b;
  	}
  	if (a === b) {
  		return true;
  	}
  	if (numberIsNaN(a) && numberIsNaN(b)) {
  		return true;
  	}
  	return false;
  };

  var polyfill$2 = function getPolyfill() {
  	return typeof Object.is === 'function' ? Object.is : implementation$1;
  };

  var shim$1 = function shimObjectIs() {
  	var polyfill = polyfill$2();
  	defineProperties_1(Object, { is: polyfill }, {
  		is: function testObjectIs() {
  			return Object.is !== polyfill;
  		}
  	});
  	return polyfill;
  };

  var polyfill$1 = callBind(polyfill$2(), Object);

  defineProperties_1(polyfill$1, {
  	getPolyfill: polyfill$2,
  	implementation: implementation$1,
  	shim: shim$1
  });

  var objectIs = polyfill$1;

  var hasToStringTag$1 = shams();



  /** @type {import('.')} */
  var fn;

  if (hasToStringTag$1) {
  	/** @type {(receiver: ThisParameterType<typeof RegExp.prototype.exec>, ...args: Parameters<typeof RegExp.prototype.exec>) => ReturnType<typeof RegExp.prototype.exec>} */
  	var $exec = callBound('RegExp.prototype.exec');
  	/** @type {object} */
  	var isRegexMarker = {};

  	var throwRegexMarker = function () {
  		throw isRegexMarker;
  	};
  	/** @type {{ toString(): never, valueOf(): never, [Symbol.toPrimitive]?(): never }} */
  	var badStringifier = {
  		toString: throwRegexMarker,
  		valueOf: throwRegexMarker
  	};

  	if (typeof Symbol.toPrimitive === 'symbol') {
  		badStringifier[Symbol.toPrimitive] = throwRegexMarker;
  	}

  	/** @type {import('.')} */
  	// @ts-expect-error TS can't figure out that the $exec call always throws
  	// eslint-disable-next-line consistent-return
  	fn = function isRegex(value) {
  		if (!value || typeof value !== 'object') {
  			return false;
  		}

  		// eslint-disable-next-line no-extra-parens
  		var descriptor = /** @type {NonNullable<typeof gOPD>} */ (gOPD$1)(/** @type {{ lastIndex?: unknown }} */ (value), 'lastIndex');
  		var hasLastIndexDataProperty = descriptor && hasOwn(descriptor, 'value');
  		if (!hasLastIndexDataProperty) {
  			return false;
  		}

  		try {
  			// eslint-disable-next-line no-extra-parens
  			$exec(value, /** @type {string} */ (/** @type {unknown} */ (badStringifier)));
  		} catch (e) {
  			return e === isRegexMarker;
  		}
  	};
  } else {
  	/** @type {(receiver: ThisParameterType<typeof Object.prototype.toString>, ...args: Parameters<typeof Object.prototype.toString>) => ReturnType<typeof Object.prototype.toString>} */
  	var $toString = callBound('Object.prototype.toString');
  	/** @const @type {'[object RegExp]'} */
  	var regexClass = '[object RegExp]';

  	/** @type {import('.')} */
  	fn = function isRegex(value) {
  		// In older browsers, typeof regex incorrectly returns 'function'
  		if (!value || (typeof value !== 'object' && typeof value !== 'function')) {
  			return false;
  		}

  		return $toString(value) === regexClass;
  	};
  }

  var isRegex = fn;

  var functionsHaveNames = function functionsHaveNames() {
  	return typeof function f() {}.name === 'string';
  };

  var gOPD = Object.getOwnPropertyDescriptor;
  if (gOPD) {
  	try {
  		gOPD([], 'length');
  	} catch (e) {
  		// IE 8 has a broken gOPD
  		gOPD = null;
  	}
  }

  functionsHaveNames.functionsHaveConfigurableNames = function functionsHaveConfigurableNames() {
  	if (!functionsHaveNames() || !gOPD) {
  		return false;
  	}
  	var desc = gOPD(function () {}, 'name');
  	return !!desc && !!desc.configurable;
  };

  var $bind = Function.prototype.bind;

  functionsHaveNames.boundFunctionsHaveNames = function boundFunctionsHaveNames() {
  	return functionsHaveNames() && typeof $bind === 'function' && function f() {}.bind().name !== '';
  };

  var functionsHaveNames_1 = functionsHaveNames;

  var hasDescriptors = require$$0();
  var functionsHaveConfigurableNames = functionsHaveNames_1.functionsHaveConfigurableNames();



  /** @type {import('.')} */
  var setFunctionName = function setFunctionName(fn, name) {
  	if (typeof fn !== 'function') {
  		throw new $TypeError$1('`fn` is not a function');
  	}
  	var loose = arguments.length > 2 && !!arguments[2];
  	if (!loose || functionsHaveConfigurableNames) {
  		if (hasDescriptors) {
  			define(/** @type {Parameters<define>[0]} */ (fn), 'name', name, true, true);
  		} else {
  			define(/** @type {Parameters<define>[0]} */ (fn), 'name', name);
  		}
  	}
  	return fn;
  };

  var $Object = Object;

  var implementation = setFunctionName(function flags() {
  	if (this == null || this !== $Object(this)) {
  		throw new $TypeError$1('RegExp.prototype.flags getter called on non-object');
  	}
  	var result = '';
  	if (this.hasIndices) {
  		result += 'd';
  	}
  	if (this.global) {
  		result += 'g';
  	}
  	if (this.ignoreCase) {
  		result += 'i';
  	}
  	if (this.multiline) {
  		result += 'm';
  	}
  	if (this.dotAll) {
  		result += 's';
  	}
  	if (this.unicode) {
  		result += 'u';
  	}
  	if (this.unicodeSets) {
  		result += 'v';
  	}
  	if (this.sticky) {
  		result += 'y';
  	}
  	return result;
  }, 'get flags', true);

  var supportsDescriptors$1 = defineProperties_1.supportsDescriptors;
  var $gOPD = Object.getOwnPropertyDescriptor;

  var polyfill = function getPolyfill() {
  	if (supportsDescriptors$1 && (/a/mig).flags === 'gim') {
  		var descriptor = $gOPD(RegExp.prototype, 'flags');
  		if (
  			descriptor
  			&& typeof descriptor.get === 'function'
  			&& 'dotAll' in RegExp.prototype
  			&& 'hasIndices' in RegExp.prototype
  		) {
  			/* eslint getter-return: 0 */
  			var calls = '';
  			var o = {};
  			Object.defineProperty(o, 'hasIndices', {
  				get: function () {
  					calls += 'd';
  				}
  			});
  			Object.defineProperty(o, 'sticky', {
  				get: function () {
  					calls += 'y';
  				}
  			});

  			descriptor.get.call(o);

  			if (calls === 'dy') {
  				return descriptor.get;
  			}
  		}
  	}
  	return implementation;
  };

  var supportsDescriptors = defineProperties_1.supportsDescriptors;


  var defineProperty = Object.defineProperty;


  var regex = /a/;

  var shim = function shimFlags() {
  	if (!supportsDescriptors || !getProto) {
  		throw new $TypeError('RegExp.prototype.flags requires a true ES5 environment that supports property descriptors');
  	}
  	var polyfill$1 = polyfill();
  	var proto = getProto(regex);
  	var descriptor = gOPD$1(proto, 'flags');
  	if (!descriptor || descriptor.get !== polyfill$1) {
  		defineProperty(proto, 'flags', {
  			configurable: true,
  			enumerable: false,
  			get: polyfill$1
  		});
  	}
  	return polyfill$1;
  };

  var flagsBound = callBind(polyfill());

  defineProperties_1(flagsBound, {
  	getPolyfill: polyfill,
  	implementation: implementation,
  	shim: shim
  });

  var regexp_prototype_flags = flagsBound;

  var getDay = callBound('Date.prototype.getDay');
  /** @type {import('.')} */
  var tryDateObject = function tryDateGetDayCall(value) {
  	try {
  		getDay(value);
  		return true;
  	} catch (e) {
  		return false;
  	}
  };

  /** @type {(value: unknown) => string} */
  var toStr = callBound('Object.prototype.toString');
  var dateClass = '[object Date]';
  var hasToStringTag = shams();

  /** @type {import('.')} */
  var isDateObject = function isDateObject(value) {
  	if (typeof value !== 'object' || value === null) {
  		return false;
  	}
  	return hasToStringTag ? tryDateObject(value) : toStr(value) === dateClass;
  };

  var isArguments = isArguments$1;

  var isDate = isDateObject;

  var getTime = Date.prototype.getTime;

  function deepEqual(actual, expected, options) {
    var opts = options || {};

    // 7.1. All identical values are equivalent, as determined by ===.
    if (opts.strict ? objectIs(actual, expected) : actual === expected) {
      return true;
    }

    // 7.3. Other pairs that do not both pass typeof value == 'object', equivalence is determined by ==.
    if (!actual || !expected || (typeof actual !== 'object' && typeof expected !== 'object')) {
      return opts.strict ? objectIs(actual, expected) : actual == expected;
    }

    /*
     * 7.4. For all other Object pairs, including Array objects, equivalence is
     * determined by having the same number of owned properties (as verified
     * with Object.prototype.hasOwnProperty.call), the same set of keys
     * (although not necessarily the same order), equivalent values for every
     * corresponding key, and an identical 'prototype' property. Note: this
     * accounts for both named and indexed properties on Arrays.
     */
    // eslint-disable-next-line no-use-before-define
    return objEquiv(actual, expected, opts);
  }

  function isUndefinedOrNull(value) {
    return value === null || value === undefined;
  }

  function isBuffer(x) {
    if (!x || typeof x !== 'object' || typeof x.length !== 'number') {
      return false;
    }
    if (typeof x.copy !== 'function' || typeof x.slice !== 'function') {
      return false;
    }
    if (x.length > 0 && typeof x[0] !== 'number') {
      return false;
    }
    return true;
  }

  function objEquiv(a, b, opts) {
    /* eslint max-statements: [2, 50] */
    var i, key;
    if (typeof a !== typeof b) { return false; }
    if (isUndefinedOrNull(a) || isUndefinedOrNull(b)) { return false; }

    // an identical 'prototype' property.
    if (a.prototype !== b.prototype) { return false; }

    if (isArguments(a) !== isArguments(b)) { return false; }

    var aIsRegex = isRegex(a);
    var bIsRegex = isRegex(b);
    if (aIsRegex !== bIsRegex) { return false; }
    if (aIsRegex || bIsRegex) {
      return a.source === b.source && regexp_prototype_flags(a) === regexp_prototype_flags(b);
    }

    if (isDate(a) && isDate(b)) {
      return getTime.call(a) === getTime.call(b);
    }

    var aIsBuffer = isBuffer(a);
    var bIsBuffer = isBuffer(b);
    if (aIsBuffer !== bIsBuffer) { return false; }
    if (aIsBuffer || bIsBuffer) { // && would work too, because both are true or both false here
      if (a.length !== b.length) { return false; }
      for (i = 0; i < a.length; i++) {
        if (a[i] !== b[i]) { return false; }
      }
      return true;
    }

    if (typeof a !== typeof b) { return false; }

    try {
      var ka = objectKeys(a);
      var kb = objectKeys(b);
    } catch (e) { // happens when one is a string literal and the other isn't
      return false;
    }
    // having the same number of owned properties (keys incorporates hasOwnProperty)
    if (ka.length !== kb.length) { return false; }

    // the same set of keys (although not necessarily the same order),
    ka.sort();
    kb.sort();
    // ~~~cheap key test
    for (i = ka.length - 1; i >= 0; i--) {
      if (ka[i] != kb[i]) { return false; }
    }
    // equivalent values for every corresponding key, and ~~~possibly expensive deep test
    for (i = ka.length - 1; i >= 0; i--) {
      key = ka[i];
      if (!deepEqual(a[key], b[key], opts)) { return false; }
    }

    return true;
  }

  var deepEqual_1 = deepEqual;

  /*
   * Date Format 1.2.3
   * (c) 2007-2009 Steven Levithan <stevenlevithan.com>
   * MIT license
   *
   * Includes enhancements by Scott Trenda <scott.trenda.net>
   * and Kris Kowal <cixar.com/~kris.kowal/>
   *
   * Accepts a date, a mask, or a date and a mask.
   * Returns a formatted version of the given date.
   * The date defaults to the current date/time.
   * The mask defaults to dateFormat.masks.default.
   */

  var dateformat = createCommonjsModule(function (module, exports) {
  (function(global) {

    var dateFormat = (function() {
        var token = /d{1,4}|m{1,4}|yy(?:yy)?|([HhMsTt])\1?|[LloSZWN]|"[^"]*"|'[^']*'/g;
        var timezone = /\b(?:[PMCEA][SDP]T|(?:Pacific|Mountain|Central|Eastern|Atlantic) (?:Standard|Daylight|Prevailing) Time|(?:GMT|UTC)(?:[-+]\d{4})?)\b/g;
        var timezoneClip = /[^-+\dA-Z]/g;
    
        // Regexes and supporting functions are cached through closure
        return function (date, mask, utc, gmt) {
    
          // You can't provide utc if you skip other args (use the 'UTC:' mask prefix)
          if (arguments.length === 1 && kindOf(date) === 'string' && !/\d/.test(date)) {
            mask = date;
            date = undefined;
          }
    
          date = date || new Date;
    
          if(!(date instanceof Date)) {
            date = new Date(date);
          }
    
          if (isNaN(date)) {
            throw TypeError('Invalid date');
          }
    
          mask = String(dateFormat.masks[mask] || mask || dateFormat.masks['default']);
    
          // Allow setting the utc/gmt argument via the mask
          var maskSlice = mask.slice(0, 4);
          if (maskSlice === 'UTC:' || maskSlice === 'GMT:') {
            mask = mask.slice(4);
            utc = true;
            if (maskSlice === 'GMT:') {
              gmt = true;
            }
          }
    
          var _ = utc ? 'getUTC' : 'get';
          var d = date[_ + 'Date']();
          var D = date[_ + 'Day']();
          var m = date[_ + 'Month']();
          var y = date[_ + 'FullYear']();
          var H = date[_ + 'Hours']();
          var M = date[_ + 'Minutes']();
          var s = date[_ + 'Seconds']();
          var L = date[_ + 'Milliseconds']();
          var o = utc ? 0 : date.getTimezoneOffset();
          var W = getWeek(date);
          var N = getDayOfWeek(date);
          var flags = {
            d:    d,
            dd:   pad(d),
            ddd:  dateFormat.i18n.dayNames[D],
            dddd: dateFormat.i18n.dayNames[D + 7],
            m:    m + 1,
            mm:   pad(m + 1),
            mmm:  dateFormat.i18n.monthNames[m],
            mmmm: dateFormat.i18n.monthNames[m + 12],
            yy:   String(y).slice(2),
            yyyy: y,
            h:    H % 12 || 12,
            hh:   pad(H % 12 || 12),
            H:    H,
            HH:   pad(H),
            M:    M,
            MM:   pad(M),
            s:    s,
            ss:   pad(s),
            l:    pad(L, 3),
            L:    pad(Math.round(L / 10)),
            t:    H < 12 ? dateFormat.i18n.timeNames[0] : dateFormat.i18n.timeNames[1],
            tt:   H < 12 ? dateFormat.i18n.timeNames[2] : dateFormat.i18n.timeNames[3],
            T:    H < 12 ? dateFormat.i18n.timeNames[4] : dateFormat.i18n.timeNames[5],
            TT:   H < 12 ? dateFormat.i18n.timeNames[6] : dateFormat.i18n.timeNames[7],
            Z:    gmt ? 'GMT' : utc ? 'UTC' : (String(date).match(timezone) || ['']).pop().replace(timezoneClip, ''),
            o:    (o > 0 ? '-' : '+') + pad(Math.floor(Math.abs(o) / 60) * 100 + Math.abs(o) % 60, 4),
            S:    ['th', 'st', 'nd', 'rd'][d % 10 > 3 ? 0 : (d % 100 - d % 10 != 10) * d % 10],
            W:    W,
            N:    N
          };
    
          return mask.replace(token, function (match) {
            if (match in flags) {
              return flags[match];
            }
            return match.slice(1, match.length - 1);
          });
        };
      })();

    dateFormat.masks = {
      'default':               'ddd mmm dd yyyy HH:MM:ss',
      'shortDate':             'm/d/yy',
      'mediumDate':            'mmm d, yyyy',
      'longDate':              'mmmm d, yyyy',
      'fullDate':              'dddd, mmmm d, yyyy',
      'shortTime':             'h:MM TT',
      'mediumTime':            'h:MM:ss TT',
      'longTime':              'h:MM:ss TT Z',
      'isoDate':               'yyyy-mm-dd',
      'isoTime':               'HH:MM:ss',
      'isoDateTime':           'yyyy-mm-dd\'T\'HH:MM:sso',
      'isoUtcDateTime':        'UTC:yyyy-mm-dd\'T\'HH:MM:ss\'Z\'',
      'expiresHeaderFormat':   'ddd, dd mmm yyyy HH:MM:ss Z'
    };

    // Internationalization strings
    dateFormat.i18n = {
      dayNames: [
        'Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat',
        'Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'
      ],
      monthNames: [
        'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec',
        'January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'
      ],
      timeNames: [
        'a', 'p', 'am', 'pm', 'A', 'P', 'AM', 'PM'
      ]
    };

  function pad(val, len) {
    val = String(val);
    len = len || 2;
    while (val.length < len) {
      val = '0' + val;
    }
    return val;
  }

  /**
   * Get the ISO 8601 week number
   * Based on comments from
   * http://techblog.procurios.nl/k/n618/news/view/33796/14863/Calculate-ISO-8601-week-and-year-in-javascript.html
   *
   * @param  {Object} `date`
   * @return {Number}
   */
  function getWeek(date) {
    // Remove time components of date
    var targetThursday = new Date(date.getFullYear(), date.getMonth(), date.getDate());

    // Change date to Thursday same week
    targetThursday.setDate(targetThursday.getDate() - ((targetThursday.getDay() + 6) % 7) + 3);

    // Take January 4th as it is always in week 1 (see ISO 8601)
    var firstThursday = new Date(targetThursday.getFullYear(), 0, 4);

    // Change date to Thursday same week
    firstThursday.setDate(firstThursday.getDate() - ((firstThursday.getDay() + 6) % 7) + 3);

    // Check if daylight-saving-time-switch occurred and correct for it
    var ds = targetThursday.getTimezoneOffset() - firstThursday.getTimezoneOffset();
    targetThursday.setHours(targetThursday.getHours() - ds);

    // Number of weeks between target Thursday and first Thursday
    var weekDiff = (targetThursday - firstThursday) / (86400000*7);
    return 1 + Math.floor(weekDiff);
  }

  /**
   * Get ISO-8601 numeric representation of the day of the week
   * 1 (for Monday) through 7 (for Sunday)
   * 
   * @param  {Object} `date`
   * @return {Number}
   */
  function getDayOfWeek(date) {
    var dow = date.getDay();
    if(dow === 0) {
      dow = 7;
    }
    return dow;
  }

  /**
   * kind-of shortcut
   * @param  {*} val
   * @return {String}
   */
  function kindOf(val) {
    if (val === null) {
      return 'null';
    }

    if (val === undefined) {
      return 'undefined';
    }

    if (typeof val !== 'object') {
      return typeof val;
    }

    if (Array.isArray(val)) {
      return 'array';
    }

    return {}.toString.call(val)
      .slice(8, -1).toLowerCase();
  }


    {
      module.exports = dateFormat;
    }
  })();
  });

  var noop = function noop() {};
  var link;
  var defaultExts = {
    extension: '',
    prefix: '',
    suffix: ''
  };

  // Alternative solution for saving files,
  // a bit slower and does not work in Safari
  // function fetchBlobFromDataURL (dataURL) {
  //   return window.fetch(dataURL).then(res => res.blob());
  // }

  var supportedEncodings = ['image/png', 'image/jpeg', 'image/webp'];
  function stream(isStart, opts) {
    if (opts === void 0) {
      opts = {};
    }
    return new Promise(function (resolve, reject) {
      opts = Object.assign({}, defaultExts, opts);
      var filename = resolveFilename(Object.assign({}, opts, {
        extension: '',
        frame: undefined
      }));
      var func = isStart ? 'streamStart' : 'streamEnd';
      var client = getClientAPI();
      if (client && client.output && typeof client[func] === 'function') {
        return client[func](Object.assign({}, opts, {
          filename: filename
        })).then(function (ev) {
          return resolve(ev);
        });
      } else {
        return resolve({
          filename: filename,
          client: false
        });
      }
    });
  }
  function streamStart(opts) {
    if (opts === void 0) {
      opts = {};
    }
    return stream(true, opts);
  }
  function streamEnd(opts) {
    if (opts === void 0) {
      opts = {};
    }
    return stream(false, opts);
  }
  function exportCanvas(canvas, opt) {
    if (opt === void 0) {
      opt = {};
    }
    var encoding = opt.encoding || 'image/png';
    if (!supportedEncodings.includes(encoding)) throw new Error("Invalid canvas encoding " + encoding);
    var extension = (encoding.split('/')[1] || '').replace(/jpeg/i, 'jpg');
    if (extension) extension = ("." + extension).toLowerCase();
    return {
      extension: extension,
      type: encoding,
      dataURL: canvas.toDataURL(encoding, opt.encodingQuality)
    };
  }
  function createBlobFromDataURL(dataURL) {
    return new Promise(function (resolve) {
      var splitIndex = dataURL.indexOf(',');
      if (splitIndex === -1) {
        resolve(new window.Blob());
        return;
      }
      var base64 = dataURL.slice(splitIndex + 1);
      var byteString = window.atob(base64);
      var type = dataURL.slice(0, splitIndex);
      var mimeMatch = /data:([^;]+)/.exec(type);
      var mime = (mimeMatch ? mimeMatch[1] : '') || undefined;
      var ab = new ArrayBuffer(byteString.length);
      var ia = new Uint8Array(ab);
      for (var i = 0; i < byteString.length; i++) {
        ia[i] = byteString.charCodeAt(i);
      }
      resolve(new window.Blob([ab], {
        type: mime
      }));
    });
  }
  function saveDataURL(dataURL, opts) {
    if (opts === void 0) {
      opts = {};
    }
    return createBlobFromDataURL(dataURL).then(function (blob) {
      return saveBlob(blob, opts);
    });
  }
  function saveBlob(blob, opts) {
    if (opts === void 0) {
      opts = {};
    }
    return new Promise(function (resolve) {
      opts = Object.assign({}, defaultExts, opts);
      var filename = opts.filename;
      var client = getClientAPI();
      if (client && typeof client.saveBlob === 'function' && client.output) {
        // native saving using a CLI tool
        return client.saveBlob(blob, Object.assign({}, opts, {
          filename: filename
        })).then(function (ev) {
          return resolve(ev);
        });
      } else {
        // force download
        if (!link) {
          link = document.createElement('a');
          link.style.visibility = 'hidden';
          link.target = '_blank';
        }
        link.download = filename;
        link.href = window.URL.createObjectURL(blob);
        document.body.appendChild(link);
        link.onclick = function () {
          link.onclick = noop;
          setTimeout(function () {
            window.URL.revokeObjectURL(blob);
            if (link.parentElement) link.parentElement.removeChild(link);
            link.removeAttribute('href');
            resolve({
              filename: filename,
              client: false
            });
          });
        };
        link.click();
      }
    });
  }
  function saveFile(data, opts) {
    if (opts === void 0) {
      opts = {};
    }
    var parts = Array.isArray(data) ? data : [data];
    var blob = new window.Blob(parts, {
      type: opts.type || ''
    });
    return saveBlob(blob, opts);
  }
  function getTimeStamp() {
    var dateFormatStr = "yyyy.mm.dd-HH.MM.ss";
    return dateformat(new Date(), dateFormatStr);
  }
  function resolveFilename(opt) {
    if (opt === void 0) {
      opt = {};
    }
    opt = Object.assign({}, opt);

    // Custom filename function
    if (typeof opt.file === 'function') {
      return opt.file(opt);
    } else if (opt.file) {
      return opt.file;
    }
    var frame = null;
    var extension = '';
    if (typeof opt.extension === 'string') extension = opt.extension;
    if (typeof opt.frame === 'number') {
      var totalFrames;
      if (typeof opt.totalFrames === 'number') {
        totalFrames = opt.totalFrames;
      } else {
        totalFrames = Math.max(10000, opt.frame);
      }
      frame = String(opt.frame).padStart(String(totalFrames).length, '0');
    }
    var layerStr = isFinite(opt.totalLayers) && isFinite(opt.layer) && opt.totalLayers > 1 ? "" + opt.layer : '';
    if (frame != null) {
      return [layerStr, frame].filter(Boolean).join('-') + extension;
    } else {
      var defaultFileName = opt.timeStamp;
      return [opt.prefix, opt.name || defaultFileName, layerStr, opt.hash, opt.suffix].filter(Boolean).join('-') + extension;
    }
  }

  // Handle some common typos
  var commonTypos = {
    dimension: 'dimensions',
    animated: 'animate',
    animating: 'animate',
    unit: 'units',
    P5: 'p5',
    pixellated: 'pixelated',
    looping: 'loop',
    pixelPerInch: 'pixels'
  };

  // Handle all other typos
  var allKeys = ['dimensions', 'units', 'pixelsPerInch', 'orientation', 'scaleToFit', 'scaleToView', 'bleed', 'pixelRatio', 'exportPixelRatio', 'maxPixelRatio', 'scaleContext', 'resizeCanvas', 'styleCanvas', 'canvas', 'context', 'attributes', 'parent', 'file', 'name', 'prefix', 'suffix', 'animate', 'playing', 'loop', 'duration', 'totalFrames', 'fps', 'playbackRate', 'timeScale', 'frame', 'time', 'flush', 'pixelated', 'hotkeys', 'p5', 'id', 'scaleToFitPadding', 'data', 'params', 'encoding', 'encodingQuality'];

  // This is fairly opinionated and forces users to use the 'data' parameter
  // if they want to pass along non-setting objects...
  var checkSettings = function checkSettings(settings) {
    var keys = Object.keys(settings);
    keys.forEach(function (key) {
      if (key in commonTypos) {
        var actual = commonTypos[key];
        console.warn("[canvas-sketch] Could not recognize the setting \"" + key + "\", did you mean \"" + actual + "\"?");
      } else if (!allKeys.includes(key)) {
        console.warn("[canvas-sketch] Could not recognize the setting \"" + key + "\"");
      }
    });
  };

  function keyboardShortcuts (opt) {
    if (opt === void 0) {
      opt = {};
    }
    var handler = function handler(ev) {
      if (!opt.enabled()) return;
      var client = getClientAPI();
      if (ev.keyCode === 83 && !ev.altKey && (ev.metaKey || ev.ctrlKey)) {
        // Cmd + S
        ev.preventDefault();
        opt.save(ev);
      } else if (ev.keyCode === 32) {
        // Space
        // TODO: what to do with this? keep it, or remove it?
        opt.togglePlay(ev);
      } else if (client && !ev.altKey && ev.keyCode === 75 && (ev.metaKey || ev.ctrlKey)) {
        // Cmd + K, only when canvas-sketch-cli is used
        ev.preventDefault();
        opt.commit(ev);
      }
    };
    var attach = function attach() {
      window.addEventListener("keydown", handler);
    };
    var detach = function detach() {
      window.removeEventListener("keydown", handler);
    };
    return {
      attach: attach,
      detach: detach
    };
  }

  var defaultUnits = 'mm';
  var data = [
  // Common Paper Sizes
  // (Mostly North-American based)
  ['postcard', 101.6, 152.4], ['poster-small', 280, 430], ['poster', 460, 610], ['poster-large', 610, 910], ['business-card', 50.8, 88.9],
  // Photographic Print Paper Sizes
  ['2r', 64, 89], ['3r', 89, 127], ['4r', 102, 152], ['5r', 127, 178],
  // 5″x7″
  ['6r', 152, 203],
  // 6″x8″
  ['8r', 203, 254],
  // 8″x10″
  ['10r', 254, 305],
  // 10″x12″
  ['11r', 279, 356],
  // 11″x14″
  ['12r', 305, 381],
  // Standard Paper Sizes
  ['a0', 841, 1189], ['a1', 594, 841], ['a2', 420, 594], ['a3', 297, 420], ['a4', 210, 297], ['a5', 148, 210], ['a6', 105, 148], ['a7', 74, 105], ['a8', 52, 74], ['a9', 37, 52], ['a10', 26, 37], ['2a0', 1189, 1682], ['4a0', 1682, 2378], ['b0', 1000, 1414], ['b1', 707, 1000], ['b1+', 720, 1020], ['b2', 500, 707], ['b2+', 520, 720], ['b3', 353, 500], ['b4', 250, 353], ['b5', 176, 250], ['b6', 125, 176], ['b7', 88, 125], ['b8', 62, 88], ['b9', 44, 62], ['b10', 31, 44], ['b11', 22, 32], ['b12', 16, 22], ['c0', 917, 1297], ['c1', 648, 917], ['c2', 458, 648], ['c3', 324, 458], ['c4', 229, 324], ['c5', 162, 229], ['c6', 114, 162], ['c7', 81, 114], ['c8', 57, 81], ['c9', 40, 57], ['c10', 28, 40], ['c11', 22, 32], ['c12', 16, 22],
  // Use inches for North American sizes,
  // as it produces less float precision errors
  ['half-letter', 5.5, 8.5, 'in'], ['letter', 8.5, 11, 'in'], ['legal', 8.5, 14, 'in'], ['junior-legal', 5, 8, 'in'], ['ledger', 11, 17, 'in'], ['tabloid', 11, 17, 'in'], ['ansi-a', 8.5, 11.0, 'in'], ['ansi-b', 11.0, 17.0, 'in'], ['ansi-c', 17.0, 22.0, 'in'], ['ansi-d', 22.0, 34.0, 'in'], ['ansi-e', 34.0, 44.0, 'in'], ['arch-a', 9, 12, 'in'], ['arch-b', 12, 18, 'in'], ['arch-c', 18, 24, 'in'], ['arch-d', 24, 36, 'in'], ['arch-e', 36, 48, 'in'], ['arch-e1', 30, 42, 'in'], ['arch-e2', 26, 38, 'in'], ['arch-e3', 27, 39, 'in']];
  var PaperSizes = data.reduce(function (dict, preset) {
    var item = {
      units: preset[3] || defaultUnits,
      dimensions: [preset[1], preset[2]]
    };
    dict[preset[0]] = item;
    dict[preset[0].replace(/-/g, ' ')] = item;
    return dict;
  }, {});

  var defined = function defined() {
  	for (var i = 0; i < arguments.length; i++) {
  		if (typeof arguments[i] !== 'undefined') {
  			return arguments[i];
  		}
  	}
  };

  var units = [ 'mm', 'cm', 'm', 'pc', 'pt', 'in', 'ft', 'px' ];

  var conversions = {
    // metric
    m: {
      system: 'metric',
      factor: 1
    },
    cm: {
      system: 'metric',
      factor: 1 / 100
    },
    mm: {
      system: 'metric',
      factor: 1 / 1000
    },
    // imperial
    pt: {
      system: 'imperial',
      factor: 1 / 72
    },
    pc: {
      system: 'imperial',
      factor: 1 / 6
    },
    in: {
      system: 'imperial',
      factor: 1
    },
    ft: {
      system: 'imperial',
      factor: 12
    }
  };

  const anchors = {
    metric: {
      unit: 'm',
      ratio: 1 / 0.0254
    },
    imperial: {
      unit: 'in',
      ratio: 0.0254
    }
  };

  function round (value, decimals) {
    return Number(Math.round(value + 'e' + decimals) + 'e-' + decimals);
  }

  function convertDistance$1 (value, fromUnit, toUnit, opts) {
    if (typeof value !== 'number' || !isFinite(value)) throw new Error('Value must be a finite number');
    if (!fromUnit || !toUnit) throw new Error('Must specify from and to units');

    opts = opts || {};
    var pixelsPerInch = defined(opts.pixelsPerInch, 96);
    var precision = opts.precision;
    var roundPixel = opts.roundPixel !== false;

    fromUnit = fromUnit.toLowerCase();
    toUnit = toUnit.toLowerCase();

    if (units.indexOf(fromUnit) === -1) throw new Error('Invalid from unit "' + fromUnit + '", must be one of: ' + units.join(', '));
    if (units.indexOf(toUnit) === -1) throw new Error('Invalid from unit "' + toUnit + '", must be one of: ' + units.join(', '));

    if (fromUnit === toUnit) {
      // We don't need to convert from A to B since they are the same already
      return value;
    }

    var toFactor = 1;
    var fromFactor = 1;
    var isToPixel = false;

    if (fromUnit === 'px') {
      fromFactor = 1 / pixelsPerInch;
      fromUnit = 'in';
    }
    if (toUnit === 'px') {
      isToPixel = true;
      toFactor = pixelsPerInch;
      toUnit = 'in';
    }

    var fromUnitData = conversions[fromUnit];
    var toUnitData = conversions[toUnit];

    // source to anchor inside source's system
    var anchor = value * fromUnitData.factor * fromFactor;

    // if systems differ, convert one to another
    if (fromUnitData.system !== toUnitData.system) {
      // regular 'm' to 'in' and so forth
      anchor *= anchors[fromUnitData.system].ratio;
    }

    var result = anchor / toUnitData.factor * toFactor;
    if (isToPixel && roundPixel) {
      result = Math.round(result);
    } else if (typeof precision === 'number' && isFinite(precision)) {
      result = round(result, precision);
    }
    return result;
  }

  var convertLength = convertDistance$1;
  var units_1 = units;
  convertLength.units = units_1;

  function getDimensionsFromPreset(dimensions, unitsTo, pixelsPerInch) {
    if (unitsTo === void 0) {
      unitsTo = 'px';
    }
    if (pixelsPerInch === void 0) {
      pixelsPerInch = 72;
    }
    if (typeof dimensions === 'string') {
      var key = dimensions.toLowerCase();
      if (!(key in PaperSizes)) {
        throw new Error("The dimension preset \"" + dimensions + "\" is not supported or could not be found; try using a4, a3, postcard, letter, etc.");
      }
      var preset = PaperSizes[key];
      return preset.dimensions.map(function (d) {
        return convertDistance(d, preset.units, unitsTo, pixelsPerInch);
      });
    } else {
      return dimensions;
    }
  }
  function convertDistance(dimension, unitsFrom, unitsTo, pixelsPerInch) {
    if (unitsFrom === void 0) {
      unitsFrom = 'px';
    }
    if (unitsTo === void 0) {
      unitsTo = 'px';
    }
    if (pixelsPerInch === void 0) {
      pixelsPerInch = 72;
    }
    return convertLength(dimension, unitsFrom, unitsTo, {
      pixelsPerInch: pixelsPerInch,
      precision: 4,
      roundPixel: true
    });
  }

  function checkIfHasDimensions(settings) {
    if (!settings.dimensions) return false;
    if (typeof settings.dimensions === "string") return true;
    if (Array.isArray(settings.dimensions) && settings.dimensions.length >= 2) return true;
    return false;
  }
  function getParentSize(props, settings) {
    // When no { dimension } is passed in node, we default to HTML canvas size
    if (!isBrowser()) {
      return [300, 150];
    }
    var element = settings.parent || window;
    if (element === window || element === document || element === document.body) {
      return [window.innerWidth, window.innerHeight];
    } else {
      var _element$getBoundingC = element.getBoundingClientRect(),
        width = _element$getBoundingC.width,
        height = _element$getBoundingC.height;
      return [width, height];
    }
  }
  function resizeCanvas(props, settings) {
    var width, height;
    var styleWidth, styleHeight;
    var canvasWidth, canvasHeight;
    var browser = isBrowser();
    var dimensions = settings.dimensions;
    var hasDimensions = checkIfHasDimensions(settings);
    var exporting = props.exporting;
    var scaleToFit = hasDimensions ? settings.scaleToFit !== false : false;
    var scaleToView = !exporting && hasDimensions ? settings.scaleToView : true;
    // in node, cancel both of these options
    if (!browser) scaleToFit = scaleToView = false;
    var units = settings.units;
    var pixelsPerInch = typeof settings.pixelsPerInch === "number" && isFinite(settings.pixelsPerInch) ? settings.pixelsPerInch : 72;
    var bleed = defined$1(settings.bleed, 0);
    var devicePixelRatio = browser ? window.devicePixelRatio : 1;
    var basePixelRatio = scaleToView ? devicePixelRatio : 1;
    var pixelRatio, exportPixelRatio;

    // If a pixel ratio is specified, we will use it.
    // Otherwise:
    //  -> If dimension is specified, use base ratio (i.e. size for export)
    //  -> If no dimension is specified, use device ratio (i.e. size for screen)
    if (typeof settings.pixelRatio === "number" && isFinite(settings.pixelRatio)) {
      // When { pixelRatio } is specified, it's also used as default exportPixelRatio.
      pixelRatio = settings.pixelRatio;
      exportPixelRatio = defined$1(settings.exportPixelRatio, pixelRatio);
    } else {
      if (hasDimensions) {
        // When a dimension is specified, use the base ratio rather than screen ratio
        pixelRatio = basePixelRatio;
        // Default to a pixel ratio of 1 so that you end up with the same dimension
        // you specified, i.e. [ 500, 500 ] is exported as 500x500 px
        exportPixelRatio = defined$1(settings.exportPixelRatio, 1);
      } else {
        // No dimension is specified, assume full-screen retina sizing
        pixelRatio = devicePixelRatio;
        // Default to screen pixel ratio, so that it's like taking a device screenshot
        exportPixelRatio = defined$1(settings.exportPixelRatio, pixelRatio);
      }
    }

    // Clamp pixel ratio
    if (typeof settings.maxPixelRatio === "number" && isFinite(settings.maxPixelRatio)) {
      pixelRatio = Math.min(settings.maxPixelRatio, pixelRatio);
    }

    // Handle export pixel ratio
    if (exporting) {
      pixelRatio = exportPixelRatio;
    }

    // parentWidth = typeof parentWidth === 'undefined' ? defaultNodeSize[0] : parentWidth;
    // parentHeight = typeof parentHeight === 'undefined' ? defaultNodeSize[1] : parentHeight;

    var _getParentSize = getParentSize(props, settings),
      parentWidth = _getParentSize[0],
      parentHeight = _getParentSize[1];
    var trimWidth, trimHeight;

    // You can specify a dimensions in pixels or cm/m/in/etc
    if (hasDimensions) {
      var result = getDimensionsFromPreset(dimensions, units, pixelsPerInch);
      var highest = Math.max(result[0], result[1]);
      var lowest = Math.min(result[0], result[1]);
      if (settings.orientation) {
        var landscape = settings.orientation === "landscape";
        width = landscape ? highest : lowest;
        height = landscape ? lowest : highest;
      } else {
        width = result[0];
        height = result[1];
      }
      trimWidth = width;
      trimHeight = height;

      // Apply bleed which is assumed to be in the same units
      width += bleed * 2;
      height += bleed * 2;
    } else {
      width = parentWidth;
      height = parentHeight;
      trimWidth = width;
      trimHeight = height;
    }

    // Real size in pixels after PPI is taken into account
    var realWidth = width;
    var realHeight = height;
    if (hasDimensions && units) {
      // Convert to digital/pixel units if necessary
      realWidth = convertDistance(width, units, "px", pixelsPerInch);
      realHeight = convertDistance(height, units, "px", pixelsPerInch);
    }

    // How big to set the 'view' of the canvas in the browser (i.e. style)
    styleWidth = Math.round(realWidth);
    styleHeight = Math.round(realHeight);

    // If we wish to scale the view to the browser window
    if (scaleToFit && !exporting && hasDimensions) {
      var aspect = width / height;
      var windowAspect = parentWidth / parentHeight;
      var scaleToFitPadding = defined$1(settings.scaleToFitPadding, 40);
      var maxWidth = Math.round(parentWidth - scaleToFitPadding * 2);
      var maxHeight = Math.round(parentHeight - scaleToFitPadding * 2);
      if (styleWidth > maxWidth || styleHeight > maxHeight) {
        if (windowAspect > aspect) {
          styleHeight = maxHeight;
          styleWidth = Math.round(styleHeight * aspect);
        } else {
          styleWidth = maxWidth;
          styleHeight = Math.round(styleWidth / aspect);
        }
      }
    }
    canvasWidth = scaleToView ? Math.round(pixelRatio * styleWidth) : Math.round(pixelRatio * realWidth);
    canvasHeight = scaleToView ? Math.round(pixelRatio * styleHeight) : Math.round(pixelRatio * realHeight);
    var viewportWidth = scaleToView ? Math.round(styleWidth) : Math.round(realWidth);
    var viewportHeight = scaleToView ? Math.round(styleHeight) : Math.round(realHeight);
    var scaleX = canvasWidth / width;
    var scaleY = canvasHeight / height;

    // Assign to current props
    return {
      bleed: bleed,
      pixelRatio: pixelRatio,
      width: width,
      height: height,
      dimensions: [width, height],
      units: units || "px",
      scaleX: scaleX,
      scaleY: scaleY,
      pixelsPerInch: pixelsPerInch,
      viewportWidth: viewportWidth,
      viewportHeight: viewportHeight,
      canvasWidth: canvasWidth,
      canvasHeight: canvasHeight,
      trimWidth: trimWidth,
      trimHeight: trimHeight,
      styleWidth: styleWidth,
      styleHeight: styleHeight
    };
  }

  var getCanvasContext_1 = getCanvasContext;
  function getCanvasContext (type, opts) {
    if (typeof type !== 'string') {
      throw new TypeError('must specify type string')
    }

    opts = opts || {};

    if (typeof document === 'undefined' && !opts.canvas) {
      return null // check for Node
    }

    var canvas = opts.canvas || document.createElement('canvas');
    if (typeof opts.width === 'number') {
      canvas.width = opts.width;
    }
    if (typeof opts.height === 'number') {
      canvas.height = opts.height;
    }

    var attribs = opts;
    var gl;
    try {
      var names = [ type ];
      // prefix GL contexts
      if (type.indexOf('webgl') === 0) {
        names.push('experimental-' + type);
      }

      for (var i = 0; i < names.length; i++) {
        gl = canvas.getContext(names[i], attribs);
        if (gl) return gl
      }
    } catch (e) {
      gl = null;
    }
    return (gl || null) // ensure null on fail
  }

  function createCanvasElement() {
    if (!isBrowser()) {
      throw new Error("It appears you are runing from Node.js or a non-browser environment. Try passing in an existing { canvas } interface instead.");
    }
    return document.createElement("canvas");
  }
  function createCanvas(settings) {
    if (settings === void 0) {
      settings = {};
    }
    var context, canvas;
    var ownsCanvas = false;
    if (settings.canvas !== false) {
      // Determine the canvas and context to create
      context = settings.context;
      var ownedCanvas;
      if (!context || typeof context === "string") {
        var newCanvas = settings.canvas;
        if (!newCanvas) {
          newCanvas = createCanvasElement();
          ownedCanvas = newCanvas;
          ownsCanvas = true;
        }
        var type = context || "2d";
        if (typeof newCanvas.getContext !== "function") {
          throw new Error("The specified { canvas } element does not have a getContext() function, maybe it is not a <canvas> tag?");
        }
        context = getCanvasContext_1(type, Object.assign({}, settings.attributes, {
          canvas: newCanvas
        }));
        if (!context) {
          throw new Error("Failed at canvas.getContext('" + type + "') - the browser may not support this context, or a different context may already be in use with this canvas.");
        }
      }
      canvas = context.canvas;

      // Some non-browser canvas implementations, e.g. Deno canvas, expose
      // context.canvas as null even when the caller supplied a valid canvas.
      if (canvas == null) {
        // just use pre-supplied canvas
        if (settings.canvas) canvas = settings.canvas;
        // or use the newly created, if { context: '2d' } for example
        else if (ownedCanvas) canvas = curCanvas;
      }

      // Ensure context matches user's canvas expectations
      if (settings.canvas && canvas !== settings.canvas) {
        throw new Error("The { canvas } and { context } settings must point to the same underlying canvas element");
      }

      // Apply pixelation to canvas if necessary, this is mostly a convenience utility
      if (settings.pixelated) {
        context.imageSmoothingEnabled = false;
        context.mozImageSmoothingEnabled = false;
        context.oImageSmoothingEnabled = false;
        context.webkitImageSmoothingEnabled = false;
        context.msImageSmoothingEnabled = false;
        canvas.style["image-rendering"] = "pixelated";
      }
    }
    return {
      canvas: canvas,
      context: context,
      ownsCanvas: ownsCanvas
    };
  }

  var SketchManager = /*#__PURE__*/function () {
    function SketchManager() {
      var _this = this;
      this._settings = {};
      this._props = {};
      this._sketch = undefined;
      this._raf = null;
      this._recordTimeout = null;

      // Some hacky things required to get around p5.js structure
      this._lastRedrawResult = undefined;
      this._isP5Resizing = false;
      this._keyboardShortcuts = keyboardShortcuts({
        enabled: function enabled() {
          return _this.settings.hotkeys !== false;
        },
        save: function save(ev) {
          if (ev.shiftKey) {
            if (_this.props.recording) {
              _this.endRecord();
              _this.run();
            } else _this.record();
          } else if (!_this.props.recording) {
            _this.exportFrame();
          }
        },
        togglePlay: function togglePlay() {
          if (_this.props.playing) _this.pause();else _this.play();
        },
        commit: function commit(ev) {
          _this.exportFrame({
            commit: true
          });
        }
      });
      this._animateHandler = function () {
        return _this.animate();
      };
      this._resizeHandler = function () {
        var changed = _this.resize();
        // Only re-render when size actually changes
        if (changed) {
          _this.render();
        }
      };
    }
    var _proto = SketchManager.prototype;
    _proto._computePlayhead = function _computePlayhead(currentTime, duration) {
      var hasDuration = typeof duration === "number" && isFinite(duration);
      return hasDuration ? currentTime / duration : 0;
    };
    _proto._computeFrame = function _computeFrame(playhead, time, totalFrames, fps) {
      return isFinite(totalFrames) && totalFrames > 1 ? Math.floor(playhead * (totalFrames - 1)) : Math.floor(fps * time);
    };
    _proto._computeCurrentFrame = function _computeCurrentFrame() {
      return this._computeFrame(this.props.playhead, this.props.time, this.props.totalFrames, this.props.fps);
    };
    _proto._getSizeProps = function _getSizeProps() {
      var props = this.props;
      return {
        width: props.width,
        height: props.height,
        pixelRatio: props.pixelRatio,
        canvasWidth: props.canvasWidth,
        canvasHeight: props.canvasHeight,
        viewportWidth: props.viewportWidth,
        viewportHeight: props.viewportHeight
      };
    };
    _proto.run = function run() {
      if (!this.sketch) throw new Error("should wait until sketch is loaded before trying to play()");

      // Start an animation frame loop if necessary
      if (this.settings.playing !== false) {
        this.play();
      }

      // Let's let this warning hang around for a few versions...
      if (typeof this.sketch.dispose === "function") {
        console.warn("In canvas-sketch@0.0.23 the dispose() event has been renamed to unload()");
      }

      // In case we aren't playing or animated, make sure we still trigger begin message...
      if (!this.props.started) {
        this._signalBegin();
        this.props.started = true;
      }

      // Render an initial frame
      this.tick();
      this.render();
      return this;
    };
    _proto._cancelTimeouts = function _cancelTimeouts() {
      if (this._raf != null && typeof window !== "undefined" && typeof window.cancelAnimationFrame === "function") {
        window.cancelAnimationFrame(this._raf);
        this._raf = null;
      }
      if (this._recordTimeout != null) {
        clearTimeout(this._recordTimeout);
        this._recordTimeout = null;
      }
    };
    _proto.play = function play() {
      var animate = this.settings.animate;
      if ("animation" in this.settings) {
        animate = true;
        console.warn("[canvas-sketch] { animation } has been renamed to { animate }");
      }
      if (!animate) return;
      if (!isBrowser()) {
        console.error("[canvas-sketch] WARN: Using { animate } in Node.js is not yet supported");
        return;
      }
      if (this.props.playing) return;
      if (!this.props.started) {
        this._signalBegin();
        this.props.started = true;
      }

      // console.log('play', this.props.time)

      // Start a render loop
      this.props.playing = true;
      this._cancelTimeouts();
      this._lastTime = browser();
      this._raf = window.requestAnimationFrame(this._animateHandler);
    };
    _proto.pause = function pause() {
      if (this.props.recording) this.endRecord();
      this.props.playing = false;
      this._cancelTimeouts();
    };
    _proto.togglePlay = function togglePlay() {
      if (this.props.playing) this.pause();else this.play();
    }

    // Stop and reset to frame zero
  ;
    _proto.stop = function stop() {
      this.pause();
      this.props.frame = 0;
      this.props.playhead = 0;
      this.props.time = 0;
      this.props.deltaTime = 0;
      this.props.started = false;
      this.render();
    };
    _proto.record = function record() {
      var _this2 = this;
      if (this.props.recording) return;
      if (!isBrowser()) {
        console.error("[canvas-sketch] WARN: Recording from Node.js is not yet supported");
        return;
      }
      this.stop();
      this.props.playing = true;
      this.props.recording = true;
      var exportOpts = this._createExportOptions({
        sequence: true
      });
      var frameInterval = 1 / this.props.fps;
      // Render each frame in the sequence
      this._cancelTimeouts();
      var _tick = function tick() {
        if (!_this2.props.recording) return Promise.resolve();
        _this2.props.deltaTime = frameInterval;
        _this2.tick();
        return _this2.exportFrame(exportOpts).then(function () {
          if (!_this2.props.recording) return; // was cancelled before
          _this2.props.deltaTime = 0;
          _this2.props.frame++;
          if (_this2.props.frame < _this2.props.totalFrames) {
            _this2.props.time += frameInterval;
            _this2.props.playhead = _this2._computePlayhead(_this2.props.time, _this2.props.duration);
            _this2._recordTimeout = setTimeout(_tick, 0);
          } else {
            console.log("Finished recording");
            _this2._signalEnd();
            _this2.endRecord();
            _this2.stop();
            _this2.run();
          }
        });
      };

      // Trigger a start event before we begin recording
      if (!this.props.started) {
        this._signalBegin();
        this.props.started = true;
      }

      // Trigger 'begin record' event
      if (this.sketch && typeof this.sketch.beginRecord === "function") {
        this._wrapContextScale(function (props) {
          return _this2.sketch.beginRecord(props);
        });
      }

      // Initiate a streaming start if necessary
      streamStart(exportOpts)["catch"](function (err) {
        console.error(err);
      }).then(function (response) {
        _this2._raf = window.requestAnimationFrame(_tick);
      });
    };
    _proto._signalBegin = function _signalBegin() {
      var _this3 = this;
      if (this.sketch && typeof this.sketch.begin === "function") {
        this._wrapContextScale(function (props) {
          return _this3.sketch.begin(props);
        });
      }
    };
    _proto._signalEnd = function _signalEnd() {
      var _this4 = this;
      if (this.sketch && typeof this.sketch.end === "function") {
        this._wrapContextScale(function (props) {
          return _this4.sketch.end(props);
        });
      }
    };
    _proto.endRecord = function endRecord() {
      var _this5 = this;
      var wasRecording = this.props.recording;
      this._cancelTimeouts();
      this.props.recording = false;
      this.props.deltaTime = 0;
      this.props.playing = false;

      // tell CLI that stream has finished
      return streamEnd()["catch"](function (err) {
        console.error(err);
      }).then(function () {
        // Trigger 'end record' event
        if (wasRecording && _this5.sketch && typeof _this5.sketch.endRecord === "function") {
          _this5._wrapContextScale(function (props) {
            return _this5.sketch.endRecord(props);
          });
        }
      });
    };
    _proto._createExportOptions = function _createExportOptions(opt) {
      if (opt === void 0) {
        opt = {};
      }
      return {
        sequence: opt.sequence,
        save: opt.save,
        fps: this.props.fps,
        frame: opt.sequence ? this.props.frame : undefined,
        file: this.settings.file,
        name: this.settings.name,
        prefix: this.settings.prefix,
        suffix: this.settings.suffix,
        encoding: this.settings.encoding,
        encodingQuality: this.settings.encodingQuality,
        timeStamp: opt.timeStamp || getTimeStamp(),
        totalFrames: isFinite(this.props.totalFrames) ? Math.max(0, this.props.totalFrames) : 1000
      };
    };
    _proto.exportFrame = function exportFrame(opt) {
      var _this6 = this;
      if (opt === void 0) {
        opt = {};
      }
      if (!this.sketch) return Promise.all([]);
      if (typeof this.sketch.preExport === "function") {
        this.sketch.preExport();
      }

      // Options for export function
      var exportOpts = this._createExportOptions(opt);
      var client = getClientAPI();
      var p = Promise.resolve();
      if (client && opt.commit && typeof client.commit === "function") {
        var commitOpts = Object.assign({}, exportOpts);
        var hash = client.commit(commitOpts);
        if (isPromise_1(hash)) p = hash;else p = Promise.resolve(hash);
      }
      return p.then(function (hash) {
        return _this6._doExportFrame(Object.assign({}, exportOpts, {
          hash: hash || ""
        }));
      }).then(function (result) {
        // Most common usecase is to export a single layer,
        // so let's optimize the user experience for that.
        if (result.length === 1) return result[0];else return result;
      });
    };
    _proto._doExportFrame = function _doExportFrame(exportOpts) {
      var _this7 = this;
      if (exportOpts === void 0) {
        exportOpts = {};
      }
      this._props.exporting = true;

      // Resize to output resolution
      this.resize();

      // Draw at this output resolution
      var drawResult = this.render();

      // The self owned canvas (may be undefined...!)
      var canvas = this.props.canvas;

      // Get list of results from render
      if (typeof drawResult === "undefined") {
        drawResult = [canvas];
      }
      drawResult = [].concat(drawResult).filter(Boolean);

      // Transform the canvas/file descriptors into a consistent format,
      // and pull out any data URLs from canvas elements
      drawResult = drawResult.map(function (result) {
        var hasDataObject = typeof result === "object" && result && ("data" in result || "dataURL" in result);
        var data = hasDataObject ? result.data : result;
        var opts = hasDataObject ? Object.assign({}, result, {
          data: data
        }) : {
          data: data
        };
        if (isCanvas(data)) {
          var encoding = opts.encoding || exportOpts.encoding;
          var encodingQuality = defined$1(opts.encodingQuality, exportOpts.encodingQuality, 0.95);
          var _exportCanvas = exportCanvas(data, {
              encoding: encoding,
              encodingQuality: encodingQuality
            }),
            dataURL = _exportCanvas.dataURL,
            extension = _exportCanvas.extension,
            type = _exportCanvas.type;
          return Object.assign(opts, {
            dataURL: dataURL,
            extension: extension,
            type: type
          });
        } else {
          return opts;
        }
      });

      // Now return to regular rendering mode
      this._props.exporting = false;
      this.resize();
      this.render();

      // And now we can save each result
      return Promise.all(drawResult.map(function (result, i, layerList) {
        // By default, if rendering multiple layers we will give them indices
        var curOpt = Object.assign({
          extension: "",
          prefix: "",
          suffix: ""
        }, exportOpts, result, {
          layer: i,
          totalLayers: layerList.length
        });

        // If export is explicitly not saving, make sure nothing saves
        // Otherwise default to the layer save option, or fallback to true
        var saveParam = exportOpts.save === false ? false : result.save;
        curOpt.save = saveParam !== false;

        // Resolve a full filename from all the options
        curOpt.filename = resolveFilename(curOpt);

        // Clean up some parameters that may be ambiguous to the user
        delete curOpt.encoding;
        delete curOpt.encodingQuality;

        // Clean it up further by just removing undefined values
        for (var k in curOpt) {
          if (typeof curOpt[k] === "undefined") delete curOpt[k];
        }
        var savePromise = Promise.resolve({});
        if (curOpt.save) {
          // Whether to actually save (download) this fragment
          var data = curOpt.data;
          if (curOpt.dataURL) {
            var dataURL = curOpt.dataURL;
            savePromise = saveDataURL(dataURL, curOpt);
          } else {
            savePromise = saveFile(data, curOpt);
          }
        }
        return savePromise.then(function (saveResult) {
          return Object.assign({}, curOpt, saveResult);
        });
      })).then(function (ev) {
        var savedEvents = ev.filter(function (e) {
          return e.save;
        });
        if (savedEvents.length > 0) {
          // Log the saved exports
          var eventWithOutput = savedEvents.find(function (e) {
            return e.outputName;
          });
          var isClient = savedEvents.some(function (e) {
            return e.client;
          });
          var isStreaming = savedEvents.some(function (e) {
            return e.stream;
          });
          var item;
          // many files, just log how many were exported
          if (savedEvents.length > 1) item = savedEvents.length;
          // in CLI, we know exact path dirname
          else if (eventWithOutput) item = eventWithOutput.outputName + "/" + savedEvents[0].filename;
          // in browser, we can only know it went to "browser download folder"
          else item = "" + savedEvents[0].filename;
          var ofSeq = "";
          if (exportOpts.sequence) {
            var hasTotalFrames = isFinite(_this7.props.totalFrames);
            ofSeq = hasTotalFrames ? " (frame " + (exportOpts.frame + 1) + " / " + _this7.props.totalFrames + ")" : " (frame " + exportOpts.frame + ")";
          } else if (savedEvents.length > 1) {
            ofSeq = " files";
          }
          var client = isClient ? "canvas-sketch-cli" : "canvas-sketch";
          var action = isStreaming ? "Streaming into" : "Exported";
          console.log("%c[" + client + "]%c " + action + " %c" + item + "%c" + ofSeq, "color: #8e8e8e;", "color: initial;", "font-weight: bold;", "font-weight: initial;");
        }
        if (typeof _this7.sketch.postExport === "function") {
          _this7.sketch.postExport();
        }
        return ev;
      });
    };
    _proto._wrapContextScale = function _wrapContextScale(cb) {
      this._preRender();
      cb(this.props);
      this._postRender();
    };
    _proto._preRender = function _preRender() {
      var props = this.props;

      // Scale context for unit sizing
      if (is2DContext(props.context) && !props.p5) {
        props.context.save();
        if (this.settings.scaleContext !== false) {
          props.context.scale(props.scaleX, props.scaleY);
        }
      } else if (props.p5) {
        props.p5.scale(props.scaleX / props.pixelRatio, props.scaleY / props.pixelRatio);
      }
    };
    _proto._postRender = function _postRender() {
      var props = this.props;
      if (is2DContext(props.context) && !props.p5) {
        props.context.restore();
      }

      // Flush by default, this may be revisited at a later point.
      // We do this to ensure toDataURL can be called immediately after.
      // Most likely browsers already handle this, so we may revisit this and
      // remove it if it improves performance without any usability issues.
      if (props.gl && this.settings.flush !== false && !props.p5) {
        props.gl.flush();
      }
    };
    _proto.tick = function tick() {
      if (this.sketch && typeof this.sketch.tick === "function") {
        this._preRender();
        this.sketch.tick(this.props);
        this._postRender();
      }
    };
    _proto.render = function render() {
      if (this.props.p5) {
        this._lastRedrawResult = undefined;
        this.props.p5.redraw();
        return this._lastRedrawResult;
      } else {
        return this.submitDrawCall();
      }
    };
    _proto.submitDrawCall = function submitDrawCall() {
      if (!this.sketch) return;
      var props = this.props;
      this._preRender();
      var drawResult;
      if (typeof this.sketch === "function") {
        drawResult = this.sketch(props);
      } else if (typeof this.sketch.render === "function") {
        drawResult = this.sketch.render(props);
      }
      this._postRender();
      return drawResult;
    };
    _proto.update = function update(opt) {
      var _this8 = this;
      if (opt === void 0) {
        opt = {};
      }
      // Currently update() is only focused on resizing,
      // but later we will support other options like switching
      // frames and such.
      var notYetSupported = ["animate"];
      Object.keys(opt).forEach(function (key) {
        if (notYetSupported.indexOf(key) >= 0) {
          throw new Error("Sorry, the { " + key + " } option is not yet supported with update().");
        }
      });
      var oldCanvas = this._settings.canvas;
      var oldContext = this._settings.context;

      // Merge new options into settings
      for (var key in opt) {
        var value = opt[key];
        if (typeof value !== "undefined") {
          // ignore undefined
          this._settings[key] = value;
        }
      }

      // Merge in time props
      var timeOpts = Object.assign({}, this._settings, opt);
      if ("time" in opt && "frame" in opt) throw new Error("You should specify { time } or { frame } but not both");else if ("time" in opt) delete timeOpts.frame;else if ("frame" in opt) delete timeOpts.time;
      if ("duration" in opt && "totalFrames" in opt) throw new Error("You should specify { duration } or { totalFrames } but not both");else if ("duration" in opt) delete timeOpts.totalFrames;else if ("totalFrames" in opt) delete timeOpts.duration;

      // Merge in user data without copying
      if ("data" in opt) this._props.data = opt.data;
      var timeProps = this.getTimeProps(timeOpts);
      Object.assign(this._props, timeProps);

      // If either canvas or context is changed, we should re-update
      if (oldCanvas !== this._settings.canvas || oldContext !== this._settings.context) {
        var _createCanvas = createCanvas(this._settings),
          canvas = _createCanvas.canvas,
          context = _createCanvas.context;
        this.props.canvas = canvas;
        this.props.context = context;

        // Delete or add a 'gl' prop for convenience
        this._setupGLKey();

        // Re-mount the new canvas if it has no parent
        this._appendCanvasIfNeeded();
      }

      // Special case to support P5.js
      if (opt.p5 && typeof opt.p5 !== "function") {
        this.props.p5 = opt.p5;
        this.props.p5.draw = function () {
          if (_this8._isP5Resizing) return;
          _this8._lastRedrawResult = _this8.submitDrawCall();
        };
      }

      // Update playing state if necessary
      if ("playing" in opt) {
        if (opt.playing) this.play();else this.pause();
      }
      checkSettings(this._settings);

      // Draw new frame
      this.resize();
      this.render();
      return this.props;
    };
    _proto.resize = function resize() {
      var oldSizes = this._getSizeProps();
      var settings = this.settings;
      var props = this.props;

      // Recompute new properties based on current setup
      var newProps = resizeCanvas(props, settings);

      // Assign to current props
      Object.assign(this._props, newProps);

      // Now we actually update the canvas width/height and style props
      var _this$props = this.props,
        pixelRatio = _this$props.pixelRatio,
        canvasWidth = _this$props.canvasWidth,
        canvasHeight = _this$props.canvasHeight,
        styleWidth = _this$props.styleWidth,
        styleHeight = _this$props.styleHeight;

      // Update canvas settings
      var canvas = this.props.canvas;
      if (canvas && settings.resizeCanvas !== false) {
        if (props.p5) {
          // P5.js specific edge case
          if (canvas.width !== canvasWidth || canvas.height !== canvasHeight) {
            this._isP5Resizing = true;
            // This causes a re-draw :\ so we ignore draws in the mean time... sorta hacky
            props.p5.pixelDensity(pixelRatio);
            props.p5.resizeCanvas(canvasWidth / pixelRatio, canvasHeight / pixelRatio, false);
            this._isP5Resizing = false;
          }
        } else {
          // Force canvas size
          if (canvas.width !== canvasWidth) canvas.width = canvasWidth;
          if (canvas.height !== canvasHeight) canvas.height = canvasHeight;
        }
        // Update canvas style
        if (isBrowser() && settings.styleCanvas !== false) {
          canvas.style.width = styleWidth + "px";
          canvas.style.height = styleHeight + "px";
        }
      }
      var newSizes = this._getSizeProps();
      var changed = !deepEqual_1(oldSizes, newSizes);
      if (changed) {
        this._sizeChanged();
      }
      return changed;
    };
    _proto._sizeChanged = function _sizeChanged() {
      // Send resize event to sketch
      if (this.sketch && typeof this.sketch.resize === "function") {
        this.sketch.resize(this.props);
      }
    };
    _proto.animate = function animate() {
      if (!this.props.playing) return;
      if (!isBrowser()) {
        console.error("[canvas-sketch] WARN: Animation in Node.js is not yet supported");
        return;
      }
      this._raf = window.requestAnimationFrame(this._animateHandler);
      var now = browser();
      var fps = this.props.fps;
      var frameIntervalMS = 1000 / fps;
      var deltaTimeMS = now - this._lastTime;
      var duration = this.props.duration;
      var hasDuration = typeof duration === "number" && isFinite(duration);
      var isNewFrame = true;
      var playbackRate = this.settings.playbackRate;
      if (playbackRate === "fixed") {
        deltaTimeMS = frameIntervalMS;
      } else if (playbackRate === "throttle") {
        if (deltaTimeMS > frameIntervalMS) {
          now = now - deltaTimeMS % frameIntervalMS;
          this._lastTime = now;
        } else {
          isNewFrame = false;
        }
      } else {
        this._lastTime = now;
      }
      var deltaTime = deltaTimeMS / 1000;
      var newTime = this.props.time + deltaTime * this.props.timeScale;

      // Handle reverse time scale
      if (newTime < 0 && hasDuration) {
        newTime = duration + newTime;
      }

      // Re-start animation
      var isFinished = false;
      var isLoopStart = false;
      var looping = this.settings.loop !== false;
      if (hasDuration && newTime >= duration) {
        // Re-start animation
        if (looping) {
          isNewFrame = true;
          newTime = newTime % duration;
          isLoopStart = true;
        } else {
          isNewFrame = false;
          newTime = duration;
          isFinished = true;
        }
        this._signalEnd();
      }
      if (isNewFrame) {
        this.props.deltaTime = deltaTime;
        this.props.time = newTime;
        this.props.playhead = this._computePlayhead(newTime, duration);
        var lastFrame = this.props.frame;
        this.props.frame = this._computeCurrentFrame();
        if (isLoopStart) this._signalBegin();
        if (lastFrame !== this.props.frame) this.tick();
        this.render();
        this.props.deltaTime = 0;
      }
      if (isFinished) {
        this.pause();
      }
    };
    _proto.dispatch = function dispatch(cb) {
      if (typeof cb !== "function") throw new Error("must pass function into dispatch()");
      cb(this.props);
      this.render();
    };
    _proto.mount = function mount() {
      this._appendCanvasIfNeeded();
    };
    _proto.unmount = function unmount() {
      if (isBrowser()) {
        window.removeEventListener("resize", this._resizeHandler);
        this._keyboardShortcuts.detach();
      }
      if (this.props.canvas.parentElement) {
        this.props.canvas.parentElement.removeChild(this.props.canvas);
      }
    };
    _proto._appendCanvasIfNeeded = function _appendCanvasIfNeeded() {
      if (!isBrowser()) return;
      if (this.settings.parent !== false && this.props.canvas && !this.props.canvas.parentElement) {
        var defaultParent = this.settings.parent || document.body;
        defaultParent.appendChild(this.props.canvas);
      }
    };
    _proto._setupGLKey = function _setupGLKey() {
      if (this.props.context) {
        if (isWebGLContext(this.props.context)) {
          this._props.gl = this.props.context;
        } else {
          delete this._props.gl;
        }
      }
    };
    _proto.getTimeProps = function getTimeProps(settings) {
      if (settings === void 0) {
        settings = {};
      }
      // Get timing data
      var duration = settings.duration;
      var totalFrames = settings.totalFrames;
      var timeScale = defined$1(settings.timeScale, 1);
      var fps = defined$1(settings.fps, 24);
      var hasDuration = typeof duration === "number" && isFinite(duration);
      var hasTotalFrames = typeof totalFrames === "number" && isFinite(totalFrames);
      var totalFramesFromDuration = hasDuration ? Math.floor(fps * duration) : undefined;
      var durationFromTotalFrames = hasTotalFrames ? totalFrames / fps : undefined;
      if (hasDuration && hasTotalFrames && totalFramesFromDuration !== totalFrames) {
        throw new Error("You should specify either duration or totalFrames, but not both. Or, they must match exactly.");
      }
      if (typeof settings.dimensions === "undefined" && typeof settings.units !== "undefined") {
        console.warn("You've specified a { units } setting but no { dimension }, so the units will be ignored.");
      }
      totalFrames = defined$1(totalFrames, totalFramesFromDuration, Infinity);
      duration = defined$1(duration, durationFromTotalFrames, Infinity);
      var startTime = settings.time;
      var startFrame = settings.frame;
      var hasStartTime = typeof startTime === "number" && isFinite(startTime);
      var hasStartFrame = typeof startFrame === "number" && isFinite(startFrame);

      // start at zero unless user specifies frame or time (but not both mismatched)
      var time = 0;
      var frame = 0;
      var playhead = 0;
      if (hasStartTime && hasStartFrame) {
        throw new Error("You should specify either start frame or time, but not both.");
      } else if (hasStartTime) {
        // User specifies time, we infer frames from FPS
        time = startTime;
        playhead = this._computePlayhead(time, duration);
        frame = this._computeFrame(playhead, time, totalFrames, fps);
      } else if (hasStartFrame) {
        // User specifies frame number, we infer time from FPS
        frame = startFrame;
        time = frame / fps;
        playhead = this._computePlayhead(time, duration);
      }
      return {
        playhead: playhead,
        time: time,
        frame: frame,
        duration: duration,
        totalFrames: totalFrames,
        fps: fps,
        timeScale: timeScale
      };
    };
    _proto.setup = function setup(settings) {
      var _this9 = this;
      if (settings === void 0) {
        settings = {};
      }
      if (this.sketch) throw new Error("Multiple setup() calls not yet supported.");
      this._settings = Object.assign({}, settings, this._settings);
      checkSettings(this._settings);

      // Get initial canvas & context
      var _createCanvas2 = createCanvas(this._settings),
        context = _createCanvas2.context,
        canvas = _createCanvas2.canvas;
      var timeProps = this.getTimeProps(this._settings);

      // Initial render state features
      this._props = _extends({}, timeProps, {
        canvas: canvas,
        context: context,
        deltaTime: 0,
        started: false,
        exporting: false,
        playing: false,
        recording: false,
        settings: this.settings,
        data: this.settings.data,
        // Export some specific actions to the sketch
        render: function render() {
          return _this9.render();
        },
        togglePlay: function togglePlay() {
          return _this9.togglePlay();
        },
        dispatch: function dispatch(cb) {
          return _this9.dispatch(cb);
        },
        tick: function tick() {
          return _this9.tick();
        },
        resize: function resize() {
          return _this9.resize();
        },
        update: function update(opt) {
          return _this9.update(opt);
        },
        exportFrame: function exportFrame(opt) {
          return _this9.exportFrame(opt);
        },
        record: function record() {
          return _this9.record();
        },
        play: function play() {
          return _this9.play();
        },
        pause: function pause() {
          return _this9.pause();
        },
        stop: function stop() {
          return _this9.stop();
        }
      });

      // For WebGL sketches, a gl variable reads a bit better
      this._setupGLKey();

      // Trigger initial resize now so that canvas is already sized
      // by the time we load the sketch
      this.resize();
    };
    _proto.loadAndRun = function loadAndRun(canvasSketch, newSettings) {
      var _this0 = this;
      return this.load(canvasSketch, newSettings).then(function () {
        _this0.run();
        return _this0;
      });
    };
    _proto.unload = function unload() {
      var _this1 = this;
      this.pause();
      if (!this.sketch) return;
      if (typeof this.sketch.unload === "function") {
        this._wrapContextScale(function (props) {
          return _this1.sketch.unload(props);
        });
      }
      this._sketch = null;
    };
    _proto.destroy = function destroy() {
      this.unload();
      this.unmount();
    };
    _proto.load = function load(createSketch, newSettings) {
      var _this10 = this;
      // User didn't specify a function
      if (typeof createSketch !== "function") {
        throw new Error("The function must take in a function as the first parameter. Example:\n  canvasSketcher(() => { ... }, settings)");
      }
      if (this.sketch) {
        this.unload();
      }
      if (typeof newSettings !== "undefined") {
        this.update(newSettings);
      }

      // This is a bit of a tricky case; we set up the auto-scaling here
      // in case the user decides to render anything to the context *before* the
      // render() function... However, users should instead use begin() function for that.
      this._preRender();
      var preload = Promise.resolve();

      // Because of P5.js's unusual structure, we have to do a bit of
      // library-specific changes to support it properly.
      if (this.settings.p5) {
        if (!isBrowser()) {
          throw new Error("[canvas-sketch] ERROR: Using p5.js in Node.js is not supported");
        }
        preload = new Promise(function (resolve) {
          var P5Constructor = _this10.settings.p5;
          var preload;
          if (P5Constructor.p5) {
            preload = P5Constructor.preload;
            P5Constructor = P5Constructor.p5;
          }

          // The sketch setup; disable loop, set sizing, etc.
          var p5Sketch = function p5Sketch(p5) {
            // Hook in preload if necessary
            if (preload) p5.preload = function () {
              return preload(p5);
            };
            p5.setup = function () {
              var props = _this10.props;
              var isGL = _this10.settings.context === "webgl";
              var renderer = isGL ? p5.WEBGL : p5.P2D;
              p5.noLoop();
              p5.pixelDensity(props.pixelRatio);
              p5.createCanvas(props.viewportWidth, props.viewportHeight, renderer);
              if (isGL && _this10.settings.attributes) {
                p5.setAttributes(_this10.settings.attributes);
              }
              _this10.update({
                p5: p5,
                canvas: p5.canvas,
                context: p5._renderer.drawingContext
              });
              resolve();
            };
          };

          // Support global and instance P5.js modes
          if (typeof P5Constructor === "function") {
            new P5Constructor(p5Sketch);
          } else {
            if (typeof window.createCanvas !== "function") {
              throw new Error("{ p5 } setting is passed but can't find p5.js in global (window) scope. Maybe you did not create it globally?\nnew p5(); // <-- attaches to global scope");
            }
            p5Sketch(window);
          }
        });
      }
      return preload.then(function () {
        // Load the user's sketch
        var loader = createSketch(_this10.props);
        if (!isPromise_1(loader)) {
          loader = Promise.resolve(loader);
        }
        return loader;
      }).then(function (sketch) {
        if (!sketch) sketch = {};
        _this10._sketch = sketch;

        // Once the sketch is loaded we can add the events
        if (isBrowser()) {
          _this10._keyboardShortcuts.attach();
          window.addEventListener("resize", _this10._resizeHandler);
        }
        _this10._postRender();

        // The initial resize() in the constructor will not have
        // triggered a resize() event on the sketch, since it was before
        // the sketch was loaded. So we send the signal here, allowing
        // users to react to the initial size before first render.
        _this10._sizeChanged();
        return _this10;
      })["catch"](function (err) {
        console.warn("Could not start sketch, the async loading function rejected with an error:\n    Error: " + err.message);
        throw err;
      });
    };
    return _createClass(SketchManager, [{
      key: "sketch",
      get: function get() {
        return this._sketch;
      }
    }, {
      key: "settings",
      get: function get() {
        return this._settings;
      }
    }, {
      key: "props",
      get: function get() {
        return this._props;
      }
    }]);
  }();

  var CACHE = 'hot-id-cache';
  var runtimeCollisions = [];
  function isHotReload() {
    var client = getClientAPI();
    return client && client.hot;
  }
  function cacheGet(id) {
    var client = getClientAPI();
    if (!client) return undefined;
    client[CACHE] = client[CACHE] || {};
    return client[CACHE][id];
  }
  function cachePut(id, data) {
    var client = getClientAPI();
    if (!client) return undefined;
    client[CACHE] = client[CACHE] || {};
    client[CACHE][id] = data;
  }
  function getTimeProp(oldManager, newSettings) {
    // Static sketches ignore the time persistency
    return newSettings.animate ? {
      time: oldManager.props.time
    } : undefined;
  }
  function canvasSketch(sketch, settings) {
    if (settings === void 0) {
      settings = {};
    }
    if (settings.p5) {
      if (settings.canvas || settings.context && typeof settings.context !== 'string') {
        throw new Error("In { p5 } mode, you can't pass your own canvas or context, unless the context is a \"webgl\" or \"2d\" string");
      }

      // Do not create a canvas on startup, since P5.js does that for us
      var context = typeof settings.context === 'string' ? settings.context : false;
      settings = Object.assign({}, settings, {
        canvas: false,
        context: context
      });
    }
    var isHot = isHotReload();
    var hotID;
    if (isHot) {
      // Use a magic name by default, force user to define each sketch if they
      // require more than one in an application. Open to other ideas on how to tackle
      // this as well...
      hotID = defined$1(settings.id, '$__DEFAULT_CANVAS_SKETCH_ID__$');
    }
    var isInjecting = isHot && typeof hotID === 'string';
    if (isInjecting && runtimeCollisions.includes(hotID)) {
      console.warn("Warning: You have multiple calls to canvasSketch() in --hot mode. You must pass unique { id } strings in settings to enable hot reload across multiple sketches. ", hotID);
      isInjecting = false;
    }
    var preload = Promise.resolve();
    if (isInjecting) {
      // Mark this as already spotted in this runtime instance
      runtimeCollisions.push(hotID);
      var previousData = cacheGet(hotID);
      if (previousData) {
        var next = function next() {
          // Grab new props from old sketch instance
          var newProps = getTimeProp(previousData.manager, settings);
          // Destroy the old instance
          previousData.manager.destroy();
          // Pass along new props
          return newProps;
        };

        // Move along the next data...
        preload = previousData.load.then(next)["catch"](next);
      }
    }
    return preload.then(function (newProps) {
      var manager = new SketchManager();
      var result;
      if (sketch) {
        // Merge with incoming data
        settings = Object.assign({}, settings, newProps);

        // Apply settings and create a canvas
        manager.setup(settings);

        // Mount to DOM
        manager.mount();

        // load the sketch first
        result = manager.loadAndRun(sketch);
      } else {
        result = Promise.resolve(manager);
      }
      if (isInjecting) {
        cachePut(hotID, {
          load: result,
          manager: manager
        });
      }
      return result;
    });
  }

  // TODO: Figure out a nice way to export things.
  canvasSketch.canvasSketch = canvasSketch;
  canvasSketch.PaperSizes = PaperSizes;

  return canvasSketch;

}));
